// @mercurial-finance/vault-sdk/dist/esm/src/vault/strategy/apricotWithoutLM.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { PublicKey, TransactionInstruction } from '@solana/web3.js';
import { TOKEN_PROGRAM_ID } from '@solana/spl-token';
import * as apricot from '@mercurial-finance/apricot-sdk';
import * as anchor from '@project-serum/anchor';
import { SEEDS } from '../constants';
export default class ApricotWithoutLMHandler {
    constructor() {
        this.address = new apricot.Addresses(apricot.PUBLIC_CONFIG);
    }
    withdraw(walletPubKey, program, strategy, vault, tokenVault, vaultState, userToken, userLp, amount, preInstructions, postInstructions, opt) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!walletPubKey)
                throw new Error('No user wallet public key');
            const [[collateralVault], [userInfoSignerPda]] = yield Promise.all([SEEDS.COLLATERAL_VAULT_PREFIX, SEEDS.APRICOT_USER_INFO_SIGNER_PREFIX].map((seed) => __awaiter(this, void 0, void 0, function* () {
                return yield PublicKey.findProgramAddress([Buffer.from(seed), new PublicKey(strategy.pubkey).toBuffer()], program.programId);
            })));
            const [basePda] = yield this.address.getBasePda();
            const [userInfo, assetPoolSpl, poolSummaries, priceSummaries, userPageStats] = yield Promise.all([
                this.address.getUserInfoKey(userInfoSignerPda),
                this.address.getAssetPoolSplKey(basePda, vaultState.tokenMint.toBase58()),
                this.address.getPoolSummariesKey(),
                this.address.getPriceSummariesKey(basePda),
                this.address.getUserPagesStatsKey(),
            ]);
            const accounts = [
                { pubkey: userInfo },
                { pubkey: assetPoolSpl },
                { pubkey: poolSummaries },
                { pubkey: priceSummaries },
                { pubkey: userInfoSignerPda },
                { pubkey: basePda },
                { pubkey: userPageStats },
            ];
            const remainingAccounts = accounts.map((account) => ({
                pubkey: account.pubkey,
                isWritable: true,
                isSigner: false,
            }));
            const txAccounts = {
                vault,
                strategy: strategy.pubkey,
                reserve: new PublicKey(strategy.state.reserve),
                strategyProgram: this.address.getProgramKey(),
                collateralVault,
                tokenVault,
                feeVault: vaultState.feeVault,
                userToken,
                userLp,
                tokenProgram: TOKEN_PROGRAM_ID,
            };
            if (opt === null || opt === void 0 ? void 0 : opt.affiliate) {
                const tx = yield opt.affiliate.affiliateProgram.methods
                    .withdrawDirectlyFromStrategy(new anchor.BN(amount), new anchor.BN(0))
                    .accounts(Object.assign(Object.assign({}, txAccounts), { partner: opt.affiliate.partner, user: opt.affiliate.user, vaultProgram: program.programId, vaultLpMint: vaultState.lpMint, owner: walletPubKey }))
                    .remainingAccounts(remainingAccounts)
                    .preInstructions(preInstructions.concat(
                // refreshUser Instruction
                new TransactionInstruction({
                    programId: this.address.getProgramKey(),
                    keys: [
                        { pubkey: userInfoSignerPda, isSigner: false, isWritable: false },
                        { pubkey: userInfo, isSigner: false, isWritable: true },
                        { pubkey: poolSummaries, isSigner: false, isWritable: false },
                    ],
                    data: Buffer.from([apricot.CMD_REFRESH_USER]),
                })))
                    .postInstructions(postInstructions)
                    .transaction();
                return tx;
            }
            const tx = yield program.methods
                .withdrawDirectlyFromStrategy(new anchor.BN(amount), new anchor.BN(0))
                .accounts(Object.assign(Object.assign({}, txAccounts), { lpMint: vaultState.lpMint, user: walletPubKey }))
                .remainingAccounts(remainingAccounts)
                .preInstructions(preInstructions.concat(
            // refreshUser Instruction
            new TransactionInstruction({
                programId: this.address.getProgramKey(),
                keys: [
                    { pubkey: userInfoSignerPda, isSigner: false, isWritable: false },
                    { pubkey: userInfo, isSigner: false, isWritable: true },
                    { pubkey: poolSummaries, isSigner: false, isWritable: false },
                ],
                data: Buffer.from([apricot.CMD_REFRESH_USER]),
            })))
                .postInstructions(postInstructions)
                .transaction();
            return tx;
        });
    }
}
