// @mercurial-finance/vault-sdk/dist/esm/src/vault/strategy/cypher.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { PublicKey } from '@solana/web3.js';
import { CypherClient, derivePublicClearingAddress, deriveAccountAddress, deriveSubAccountAddress, derivePoolNodeAddress, derivePoolNodeVaultAddress, derivePoolNodeVaultSigner, CONFIGS, } from '@mercurial-finance/cypher-client';
import { TOKEN_PROGRAM_ID } from '@solana/spl-token';
import BN from 'bn.js';
import { SEEDS } from '../constants';
import { getOrCreateATAInstruction } from '../utils';
export default class CypherHandler {
    constructor(cluster, program) {
        this.cluster = cluster;
        this.program = program;
        this.cypherClient = new CypherClient(cluster, program.provider.connection.rpcEndpoint);
    }
    withdraw(walletPubKey, program, strategy, vault, tokenVault, vaultState, userToken, userLp, amount, preInstructions, postInstructions, opt) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!walletPubKey)
                throw new Error('No user wallet public key');
            const strategyBuffer = new PublicKey(strategy.pubkey).toBuffer();
            const [collateralVault] = yield PublicKey.findProgramAddress([Buffer.from(SEEDS.COLLATERAL_VAULT_PREFIX), strategyBuffer], program.programId);
            const [strategyOwnerPubkey] = yield PublicKey.findProgramAddress([Buffer.from(SEEDS.CYPHER), strategyBuffer], program.programId);
            const [strategyOwnerATA] = yield getOrCreateATAInstruction(vaultState.tokenMint, strategyOwnerPubkey, this.program.provider.connection);
            const [clearingPubkey] = derivePublicClearingAddress(this.cypherClient.cypherPID);
            const [masterAccPubkey] = deriveAccountAddress(strategyOwnerPubkey, 0, this.cypherClient.cypherPID);
            const [subAccPubkey] = deriveSubAccountAddress(masterAccPubkey, 0, this.cypherClient.cypherPID);
            const [poolNodePubKey] = derivePoolNodeAddress(strategy.state.reserve, 0, this.cypherClient.cypherPID);
            const [poolTokenVaultPubkey] = derivePoolNodeVaultAddress(poolNodePubKey, this.cypherClient.cypherPID);
            const [vaultSigner] = derivePoolNodeVaultSigner(poolNodePubKey, this.cypherClient.cypherPID);
            const accounts = [
                { pubkey: clearingPubkey },
                { pubkey: CONFIGS[this.cluster].CACHE },
                { pubkey: masterAccPubkey, isWritable: true },
                { pubkey: subAccPubkey, isWritable: true },
                { pubkey: poolNodePubKey, isWritable: true },
                { pubkey: poolTokenVaultPubkey, isWritable: true },
                { pubkey: vaultState.tokenMint },
                { pubkey: strategyOwnerPubkey, isWritable: true },
                { pubkey: strategyOwnerATA, isWritable: true },
                { pubkey: vaultSigner, isWritable: true },
            ];
            const remainingAccounts = [];
            for (const account of accounts) {
                remainingAccounts.push({
                    pubkey: account.pubkey,
                    isWritable: !!account.isWritable,
                    isSigner: false,
                });
            }
            const txAccounts = {
                vault,
                strategy: new PublicKey(strategy.pubkey),
                reserve: new PublicKey(strategy.state.reserve),
                strategyProgram: this.cypherClient.cypherPID,
                collateralVault,
                feeVault: vaultState.feeVault,
                tokenVault,
                userToken,
                userLp,
                tokenProgram: TOKEN_PROGRAM_ID,
            };
            if (opt === null || opt === void 0 ? void 0 : opt.affiliate) {
                const tx = yield opt.affiliate.affiliateProgram.methods
                    .withdrawDirectlyFromStrategy(new BN(amount), new BN(0))
                    .accounts(Object.assign(Object.assign({}, txAccounts), { partner: opt.affiliate.partner, user: opt.affiliate.user, vaultProgram: program.programId, vaultLpMint: vaultState.lpMint, owner: walletPubKey }))
                    .remainingAccounts(remainingAccounts)
                    .preInstructions(preInstructions)
                    .postInstructions(postInstructions)
                    .transaction();
                return tx;
            }
            const tx = yield program.methods
                .withdrawDirectlyFromStrategy(new BN(amount), new BN(0))
                .accounts(Object.assign(Object.assign({}, txAccounts), { lpMint: vaultState.lpMint, user: walletPubKey }))
                .remainingAccounts(remainingAccounts)
                .preInstructions(preInstructions)
                .postInstructions(postInstructions)
                .transaction();
            return tx;
        });
    }
}
