// @mercurial-finance/vault-sdk/dist/esm/src/vault/strategy/drift.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { Keypair, PublicKey } from '@solana/web3.js';
import { DriftClient, initialize, MainnetSpotMarkets, DevnetSpotMarkets, getUserStatsAccountPublicKey, getUserAccountPublicKeySync, getDriftStateAccountPublicKey, getDriftSignerPublicKey, getSpotMarketVaultPublicKey, Wallet, } from '@mercurial-finance/drift-sdk';
import { TOKEN_PROGRAM_ID } from '@solana/spl-token';
import BN from 'bn.js';
import { SEEDS } from '../constants';
export default class DriftHandler {
    constructor(cluster, program) {
        this.sdkConfig = initialize({ env: cluster === 'testnet' ? 'mainnet-beta' : cluster });
        this.cluster = cluster;
        this.driftClient = new DriftClient({
            connection: program.provider.connection,
            programID: new PublicKey(this.sdkConfig.DRIFT_PROGRAM_ID),
            wallet: new Wallet(Keypair.generate()),
        });
        this.programId = new PublicKey(this.sdkConfig.DRIFT_PROGRAM_ID);
    }
    withdraw(walletPubKey, program, strategy, vault, tokenVault, vaultState, userToken, userLp, amount, preInstructions, postInstructions, opt) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!walletPubKey)
                throw new Error('No user wallet public key');
            yield this.driftClient.subscribe();
            const spotMarkets = this.cluster === 'devnet' ? DevnetSpotMarkets : MainnetSpotMarkets;
            const spotMarket = spotMarkets.find((market) => market.mint.equals(vaultState.tokenMint));
            if (!spotMarket)
                throw new Error('Spot market not found');
            yield this.driftClient.accountSubscriber.addSpotMarket(spotMarket.marketIndex);
            const strategyBuffer = new PublicKey(strategy.pubkey).toBuffer();
            const [collateralVault] = yield PublicKey.findProgramAddress([Buffer.from(SEEDS.COLLATERAL_VAULT_PREFIX), strategyBuffer], program.programId);
            // subAccountId is 16 bytes, therefore we slice 2 bytes only
            const subAccountId = new BN(strategy.state.bumps.slice(0, 2), 'le');
            const userPubKey = getUserAccountPublicKeySync(this.programId, vault, subAccountId.toNumber());
            const userStatsPubKey = getUserStatsAccountPublicKey(this.programId, vault);
            const spotMarketVaultPubKey = yield getSpotMarketVaultPublicKey(this.programId, spotMarket.marketIndex);
            const driftState = yield getDriftStateAccountPublicKey(this.programId);
            const driftSigner = getDriftSignerPublicKey(this.programId);
            const accounts = [
                { pubkey: userPubKey, isWritable: true },
                { pubkey: userStatsPubKey, isWritable: true },
                { pubkey: spotMarketVaultPubKey, isWritable: true },
                { pubkey: driftState },
                { pubkey: driftSigner },
            ];
            const remainingAccounts = [];
            for (const account of accounts) {
                remainingAccounts.push({
                    pubkey: account.pubkey,
                    isWritable: !!account.isWritable,
                    isSigner: false,
                });
            }
            const driftRemainingAccounts = this.driftClient.getRemainingAccounts({
                userAccounts: [],
                writableSpotMarketIndexes: [spotMarket.marketIndex],
                useMarketLastSlotCache: false,
            });
            remainingAccounts.push(...driftRemainingAccounts);
            // prevent duplicate as spot market account pubkey will be add on program side
            const remainingAccountsWithoutReserve = remainingAccounts.filter(({ pubkey }) => !pubkey.equals(strategy.state.reserve));
            const txAccounts = {
                vault,
                strategy: new PublicKey(strategy.pubkey),
                reserve: new PublicKey(strategy.state.reserve),
                strategyProgram: this.sdkConfig.DRIFT_PROGRAM_ID,
                collateralVault,
                feeVault: vaultState.feeVault,
                tokenVault,
                userToken,
                userLp,
                tokenProgram: TOKEN_PROGRAM_ID,
            };
            if (opt === null || opt === void 0 ? void 0 : opt.affiliate) {
                const tx = yield opt.affiliate.affiliateProgram.methods
                    .withdrawDirectlyFromStrategy(new BN(amount), new BN(0))
                    .accounts(Object.assign(Object.assign({}, txAccounts), { partner: opt.affiliate.partner, user: opt.affiliate.user, vaultProgram: program.programId, vaultLpMint: vaultState.lpMint, owner: walletPubKey }))
                    .remainingAccounts(remainingAccountsWithoutReserve)
                    .preInstructions(preInstructions)
                    .postInstructions(postInstructions)
                    .transaction();
                return tx;
            }
            const tx = yield program.methods
                .withdrawDirectlyFromStrategy(new BN(amount), new BN(0))
                .accounts(Object.assign(Object.assign({}, txAccounts), { lpMint: vaultState.lpMint, user: walletPubKey }))
                .remainingAccounts(remainingAccountsWithoutReserve)
                .preInstructions(preInstructions)
                .postInstructions(postInstructions)
                .transaction();
            return tx;
        });
    }
}
