// @mercurial-finance/vault-sdk/dist/esm/src/vault/strategy/frakt.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { PublicKey, SYSVAR_CLOCK_PUBKEY, SystemProgram, } from '@solana/web3.js';
import { TOKEN_PROGRAM_ID } from '@solana/spl-token';
import BN from 'bn.js';
import { SEEDS } from '../constants';
import { getOrCreateATAInstruction } from '../utils';
const FRAKT_PROGRAM_ID = new PublicKey('A66HabVL3DzNzeJgcHYtRRNW1ZRMKwBfrdSR4kLsZ9DJ');
const FRAKT_ADMIN_FEE_PUBKEY = new PublicKey('9aTtUqAnuSMndCpjcPosRNf3fCkrTQAV8C8GERf3tZi3');
export default class FraktHandler {
    constructor(program) {
        this.connection = program.provider.connection;
    }
    withdraw(walletPubKey, program, strategy, vault, tokenVault, vaultState, userToken, userLp, amount, preInstructions, postInstructions, opt) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!walletPubKey)
                throw new Error('No user wallet public key');
            const [liqOwner] = yield PublicKey.findProgramAddress([Buffer.from(SEEDS.FRAKT_LENDING), strategy.state.reserve.toBuffer()], FRAKT_PROGRAM_ID);
            const strategyBuffer = new PublicKey(strategy.pubkey).toBuffer();
            const strategyReserveBuffer = new PublicKey(strategy.state.reserve).toBuffer();
            const [collateralVault] = yield PublicKey.findProgramAddress([Buffer.from(SEEDS.COLLATERAL_VAULT_PREFIX), strategyBuffer], program.programId);
            const [strategyOwner] = yield PublicKey.findProgramAddress([Buffer.from(SEEDS.FRAKT), strategyBuffer], program.programId);
            const [tokenAccount, createTokenAccountIx] = yield getOrCreateATAInstruction(vaultState.tokenMint, strategyOwner, this.connection, {
                payer: walletPubKey,
            });
            createTokenAccountIx && preInstructions.push(createTokenAccountIx);
            const [deposit] = yield PublicKey.findProgramAddress([Buffer.from(SEEDS.DEPOSIT), strategyReserveBuffer, new PublicKey(strategyOwner).toBuffer()], FRAKT_PROGRAM_ID);
            const accounts = [
                { pubkey: strategyOwner, isWritable: true },
                { pubkey: tokenAccount, isWritable: true },
                { pubkey: liqOwner, isWritable: true },
                { pubkey: deposit, isWritable: true },
                { pubkey: FRAKT_ADMIN_FEE_PUBKEY, isWritable: true },
                { pubkey: SystemProgram.programId },
                { pubkey: SYSVAR_CLOCK_PUBKEY },
            ];
            const remainingAccounts = [];
            for (const account of accounts) {
                remainingAccounts.push({
                    pubkey: account.pubkey,
                    isWritable: !!account.isWritable,
                    isSigner: false,
                });
            }
            // prevent duplicate as spot market account pubkey will be add on program side
            const remainingAccountsWithoutReserve = remainingAccounts.filter(({ pubkey }) => !pubkey.equals(strategy.state.reserve));
            const txAccounts = {
                vault,
                strategy: new PublicKey(strategy.pubkey),
                reserve: new PublicKey(strategy.state.reserve),
                strategyProgram: FRAKT_PROGRAM_ID,
                collateralVault,
                feeVault: vaultState.feeVault,
                tokenVault,
                userToken,
                userLp,
                tokenProgram: TOKEN_PROGRAM_ID,
            };
            if (opt === null || opt === void 0 ? void 0 : opt.affiliate) {
                const tx = yield opt.affiliate.affiliateProgram.methods
                    .withdrawDirectlyFromStrategy(new BN(amount), new BN(0))
                    .accounts(Object.assign(Object.assign({}, txAccounts), { partner: opt.affiliate.partner, user: opt.affiliate.user, vaultProgram: program.programId, vaultLpMint: vaultState.lpMint, owner: walletPubKey }))
                    .remainingAccounts(remainingAccountsWithoutReserve)
                    .preInstructions(preInstructions)
                    .postInstructions(postInstructions)
                    .transaction();
                return tx;
            }
            const tx = yield program.methods
                .withdrawDirectlyFromStrategy(new BN(amount), new BN(0))
                .accounts(Object.assign(Object.assign({}, txAccounts), { lpMint: vaultState.lpMint, user: walletPubKey }))
                .remainingAccounts(remainingAccountsWithoutReserve)
                .preInstructions(preInstructions)
                .postInstructions(postInstructions)
                .transaction();
            return tx;
        });
    }
}
