// @mercurial-finance/vault-sdk/dist/esm/src/vault/strategy/portWithoutLM.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { PublicKey, SYSVAR_CLOCK_PUBKEY } from '@solana/web3.js';
import { TOKEN_PROGRAM_ID } from '@solana/spl-token';
import * as anchor from '@project-serum/anchor';
import * as port from '@mercurial-finance/port-sdk';
import { SEEDS } from '../constants';
export default class PortWithoutLMHandler {
    constructor(strategyProgram) {
        this.strategyProgram = strategyProgram;
    }
    getReserveState(program, reserve) {
        return __awaiter(this, void 0, void 0, function* () {
            const account = yield program.provider.connection.getAccountInfo(reserve);
            const state = port.ReserveLayout.decode(account.data);
            return {
                collateral: {
                    mintPubkey: state.collateral.mintPubkey,
                    mintTotalSupply: Number(state.collateral.mintTotalSupply.toU64()),
                    supplyPubkey: state.collateral.supplyPubkey.toString(),
                },
                state,
            };
        });
    }
    withdraw(walletPubKey, program, strategy, vault, tokenVault, vaultState, userToken, userLp, amount, preInstructions, postInstructions, opt) {
        return __awaiter(this, void 0, void 0, function* () {
            const { state } = yield this.getReserveState(program, strategy.state.reserve);
            const strategyBuffer = new PublicKey(strategy.pubkey).toBuffer();
            let [collateralVault] = yield PublicKey.findProgramAddress([Buffer.from(SEEDS.COLLATERAL_VAULT_PREFIX), strategyBuffer], program.programId);
            const { collateral: portCollateral, lendingMarket, liquidity } = state;
            const oraclePubkey = liquidity.oracleOption === 0 ? null : liquidity.oraclePubkey;
            const [lendingMarketAuthority] = yield PublicKey.findProgramAddress([lendingMarket.toBuffer()], this.strategyProgram);
            const accountData = [
                { pubkey: liquidity.supplyPubkey, isWritable: true },
                { pubkey: lendingMarket },
                { pubkey: lendingMarketAuthority },
                { pubkey: portCollateral.mintPubkey, isWritable: true },
                { pubkey: SYSVAR_CLOCK_PUBKEY },
            ];
            const remainingAccounts = [];
            for (const account of accountData) {
                remainingAccounts.push({
                    pubkey: account.pubkey,
                    isWritable: !!account.isWritable,
                    isSigner: false,
                });
            }
            const txAccounts = {
                vault,
                strategy: strategy.pubkey,
                reserve: strategy.state.reserve,
                strategyProgram: this.strategyProgram,
                collateralVault,
                tokenVault,
                feeVault: vaultState.feeVault,
                userToken,
                userLp,
                tokenProgram: TOKEN_PROGRAM_ID,
            };
            if (opt === null || opt === void 0 ? void 0 : opt.affiliate) {
                const tx = yield opt.affiliate.affiliateProgram.methods
                    .withdrawDirectlyFromStrategy(new anchor.BN(amount), new anchor.BN(0))
                    .accounts(Object.assign(Object.assign({}, txAccounts), { partner: opt.affiliate.partner, user: opt.affiliate.user, vaultProgram: program.programId, vaultLpMint: vaultState.lpMint, owner: walletPubKey }))
                    .remainingAccounts(remainingAccounts)
                    .preInstructions(preInstructions.concat([
                    port.refreshReserveInstruction(strategy.state.reserve, oraclePubkey, this.strategyProgram),
                ]))
                    .postInstructions(postInstructions)
                    .transaction();
                return tx;
            }
            const tx = yield program.methods
                .withdrawDirectlyFromStrategy(amount, new anchor.BN(0))
                .accounts(Object.assign(Object.assign({}, txAccounts), { lpMint: vaultState.lpMint, user: walletPubKey }))
                .remainingAccounts(remainingAccounts)
                .preInstructions(preInstructions.concat([
                port.refreshReserveInstruction(strategy.state.reserve, oraclePubkey, this.strategyProgram),
            ]))
                .postInstructions(postInstructions)
                .transaction();
            return tx;
        });
    }
}
