// @mercurial-finance/vault-sdk/dist/esm/src/vault/strategy/psylend.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { PublicKey, } from '@solana/web3.js';
import { IDL as PsyLendIDL, keys, instructions } from '@mithraic-labs/psylend-utils';
import { TOKEN_PROGRAM_ID } from '@solana/spl-token';
import { BN, Program, AnchorProvider } from '@project-serum/anchor';
import { SEEDS } from '../constants';
import { getOrCreateATAInstruction } from '../utils';
export default class PsyLendHandler {
    constructor(cluster, program) {
        this.cluster = cluster;
        this.connection = program.provider.connection;
        this.psyLendProgram = new Program(PsyLendIDL, keys.psyLendMainnetProgramKey, new AnchorProvider(program.provider.connection, {}, AnchorProvider.defaultOptions()));
    }
    withdraw(walletPubKey, program, strategy, vault, tokenVault, vaultState, userToken, userLp, amount, preInstructions, postInstructions, opt) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!walletPubKey)
                throw new Error('No user wallet public key');
            const strategyBuffer = new PublicKey(strategy.pubkey).toBuffer();
            const [collateralVault] = PublicKey.findProgramAddressSync([Buffer.from(SEEDS.COLLATERAL_VAULT_PREFIX), strategyBuffer], program.programId);
            const { reserve } = strategy.state;
            const reserveState = yield this.psyLendProgram.account.reserve.fetchNullable(reserve);
            if (!reserveState)
                throw new Error('No user reserve account');
            const { market, depositNoteMint, pythOraclePrice, feeNoteVault, vault: reserveVault } = reserveState;
            const marketState = yield this.psyLendProgram.account.market.fetchNullable(market);
            if (!marketState)
                throw new Error('No user reserve market account');
            const { marketAuthority } = marketState;
            const [strategyOwnerPubkey] = PublicKey.findProgramAddressSync([Buffer.from(SEEDS.PSYLEND), strategyBuffer], program.programId);
            const [strategyOwnerAccount] = PublicKey.findProgramAddressSync([Buffer.from(SEEDS.PSYLEND_OWNER), reserve.toBuffer(), strategyOwnerPubkey.toBuffer()], keys.psyLendMainnetProgramKey);
            const [tokenAccount, createTokenAccountIx] = yield getOrCreateATAInstruction(vaultState.tokenMint, strategyOwnerPubkey, this.connection, {
                payer: walletPubKey,
            });
            createTokenAccountIx && preInstructions.push(createTokenAccountIx);
            const accounts = [
                { pubkey: market, isWritable: true },
                { pubkey: marketAuthority, isWritable: true },
                { pubkey: strategyOwnerAccount, isWritable: true },
                { pubkey: strategyOwnerPubkey, isWritable: true },
                { pubkey: reserveVault, isWritable: true },
                { pubkey: depositNoteMint, isWritable: true },
                { pubkey: tokenAccount, isWritable: true },
            ];
            const remainingAccounts = [];
            for (const account of accounts) {
                remainingAccounts.push({
                    pubkey: account.pubkey,
                    isWritable: true,
                    isSigner: false,
                });
            }
            const txAccounts = {
                vault,
                strategy: new PublicKey(strategy.pubkey),
                reserve: new PublicKey(strategy.state.reserve),
                strategyProgram: keys.psyLendMainnetProgramKey,
                collateralVault,
                feeVault: vaultState.feeVault,
                tokenVault,
                userToken,
                userLp,
                tokenProgram: TOKEN_PROGRAM_ID,
            };
            const accrueInterestIx = yield instructions.accrueInterestIx(this.psyLendProgram, market, marketAuthority, reserve, feeNoteVault, depositNoteMint);
            preInstructions.push(accrueInterestIx);
            const refreshReserveIx = yield instructions.refreshReserveIx(this.psyLendProgram, market, reserve, pythOraclePrice);
            preInstructions.push(refreshReserveIx);
            if (opt === null || opt === void 0 ? void 0 : opt.affiliate) {
                const tx = yield opt.affiliate.affiliateProgram.methods
                    .withdrawDirectlyFromStrategy(new BN(amount), new BN(0))
                    .accounts(Object.assign(Object.assign({}, txAccounts), { partner: opt.affiliate.partner, user: opt.affiliate.user, vaultProgram: program.programId, vaultLpMint: vaultState.lpMint, owner: walletPubKey }))
                    .remainingAccounts(remainingAccounts)
                    .preInstructions(preInstructions)
                    .postInstructions(postInstructions)
                    .transaction();
                return tx;
            }
            const tx = yield program.methods
                .withdrawDirectlyFromStrategy(new BN(amount), new BN(0))
                .accounts(Object.assign(Object.assign({}, txAccounts), { lpMint: vaultState.lpMint, user: walletPubKey }))
                .remainingAccounts(remainingAccounts)
                .preInstructions(preInstructions)
                .postInstructions(postInstructions)
                .transaction();
            return tx;
        });
    }
}
