// @mercurial-finance/vault-sdk/dist/esm/src/vault/strategy/tulip.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { PublicKey, SYSVAR_CLOCK_PUBKEY } from '@solana/web3.js';
import { TOKEN_PROGRAM_ID } from '@solana/spl-token';
import * as tulip from '@mercurial-finance/tulip-platform-sdk';
import * as anchor from '@project-serum/anchor';
import { SEEDS } from '../constants';
export default class TulipHandler {
    withdraw(walletPubKey, program, strategy, vault, tokenVault, vaultState, userToken, userLp, amount, preInstructions, postInstructions, opt) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!walletPubKey)
                throw new Error('No user wallet public key');
            const lendingPools = tulip.LENDING_RESERVES;
            const lendingPool = Object.values(lendingPools).find((lendingPool) => lendingPool.account === strategy.state.reserve.toBase58());
            if (!lendingPool)
                throw new Error('Cannot find tulip lending pool');
            const collateralMint = lendingPool.collateralTokenMint;
            const strategyBuffer = new PublicKey(strategy.pubkey).toBuffer();
            const [collateralVault] = yield PublicKey.findProgramAddress([Buffer.from(SEEDS.COLLATERAL_VAULT_PREFIX), strategyBuffer], program.programId);
            const [derivedLendingMarketAuthority] = yield anchor.web3.PublicKey.findProgramAddress([new anchor.web3.PublicKey(tulip.getLendingMarketAccount()).toBytes()], tulip.LENDING_PROGRAM_ID);
            const accounts = [
                { pubkey: lendingPool.liquiditySupplyTokenAccount, isWritable: true },
                { pubkey: tulip.getLendingMarketAccount(), isWritable: true },
                { pubkey: derivedLendingMarketAuthority },
                { pubkey: collateralMint, isWritable: true },
                { pubkey: SYSVAR_CLOCK_PUBKEY },
            ];
            const remainingAccounts = [];
            for (const account of accounts) {
                remainingAccounts.push({
                    pubkey: new PublicKey(account.pubkey),
                    isWritable: !!account.isWritable,
                    isSigner: false,
                });
            }
            const txAccounts = {
                vault,
                strategy: new PublicKey(strategy.pubkey),
                reserve: new PublicKey(strategy.state.reserve),
                strategyProgram: tulip.LENDING_PROGRAM_ID,
                collateralVault,
                feeVault: vaultState.feeVault,
                tokenVault,
                userToken,
                userLp,
                tokenProgram: TOKEN_PROGRAM_ID,
            };
            if (opt === null || opt === void 0 ? void 0 : opt.affiliate) {
                const tx = yield opt.affiliate.affiliateProgram.methods
                    .withdrawDirectlyFromStrategy(new anchor.BN(amount), new anchor.BN(0))
                    .accounts(Object.assign(Object.assign({}, txAccounts), { partner: opt.affiliate.partner, user: opt.affiliate.user, vaultProgram: program.programId, vaultLpMint: vaultState.lpMint, owner: walletPubKey }))
                    .remainingAccounts(remainingAccounts)
                    .preInstructions(preInstructions.concat([
                    tulip.refreshReserve({
                        reserveAccount: strategy.state.reserve,
                        priceAccount: new PublicKey(tulip.getPriceFeedsForReserve(lendingPool.name).price_account),
                    }),
                ]))
                    .postInstructions(postInstructions)
                    .transaction();
                return tx;
            }
            const tx = yield program.methods
                .withdrawDirectlyFromStrategy(new anchor.BN(amount), new anchor.BN(0))
                .accounts(Object.assign(Object.assign({}, txAccounts), { lpMint: vaultState.lpMint, user: walletPubKey }))
                .remainingAccounts(remainingAccounts)
                .preInstructions(preInstructions.concat([
                tulip.refreshReserve({
                    reserveAccount: strategy.state.reserve,
                    priceAccount: new PublicKey(tulip.getPriceFeedsForReserve(lendingPool.name).price_account),
                }),
            ]))
                .postInstructions(postInstructions)
                .transaction();
            return tx;
        });
    }
}
