// @quarryprotocol/quarry-sdk/dist/esm/wrappers/operator/index.js
import { __awaiter } from "tslib";
import { TransactionEnvelope } from "@saberhq/solana-contrib";
import { u64 } from "@saberhq/token-utils";
import { Keypair, SystemProgram } from "@solana/web3.js";
import { findQuarryAddress } from "..";
import { findOperatorAddress } from "./pda";
/**
 * Operator helper functions.
 */
export class Operator {
    constructor(sdk, key, data) {
        this.sdk = sdk;
        this.key = key;
        this.data = data;
    }
    get program() {
        return this.sdk.programs.Operator;
    }
    /**
     * Reloads the Operator's data.
     * @returns
     */
    reload() {
        return __awaiter(this, void 0, void 0, function* () {
            const data = yield this.program.account.operator.fetch(this.key);
            return new Operator(this.sdk, this.key, data);
        });
    }
    static load({ sdk, key, }) {
        return __awaiter(this, void 0, void 0, function* () {
            const program = sdk.programs.Operator;
            const data = (yield program.account.operator.fetchNullable(key));
            if (!data) {
                return null;
            }
            return new Operator(sdk, key, data);
        });
    }
    static createOperator({ sdk, rewarder, baseKP = Keypair.generate(), admin = sdk.provider.wallet.publicKey, payer = sdk.provider.wallet.publicKey, }) {
        return __awaiter(this, void 0, void 0, function* () {
            const [operatorKey] = yield findOperatorAddress(baseKP.publicKey, sdk.programs.Operator.programId);
            return {
                key: operatorKey,
                tx: new TransactionEnvelope(sdk.provider, [
                    sdk.programs.Operator.instruction.createOperatorV2({
                        accounts: {
                            base: baseKP.publicKey,
                            operator: operatorKey,
                            rewarder,
                            admin,
                            payer,
                            systemProgram: SystemProgram.programId,
                            quarryMineProgram: sdk.programs.Mine.programId,
                        },
                    }),
                ], [baseKP]),
            };
        });
    }
    setAdmin(delegate) {
        return new TransactionEnvelope(this.sdk.provider, [
            this.program.instruction.setAdmin({
                accounts: {
                    operator: this.key,
                    admin: this.sdk.provider.wallet.publicKey,
                    delegate,
                },
            }),
        ]);
    }
    setRateSetter(delegate) {
        return new TransactionEnvelope(this.sdk.provider, [
            this.program.instruction.setRateSetter({
                accounts: {
                    operator: this.key,
                    admin: this.sdk.provider.wallet.publicKey,
                    delegate,
                },
            }),
        ]);
    }
    setQuarryCreator(delegate) {
        return new TransactionEnvelope(this.sdk.provider, [
            this.program.instruction.setQuarryCreator({
                accounts: {
                    operator: this.key,
                    admin: this.sdk.provider.wallet.publicKey,
                    delegate,
                },
            }),
        ]);
    }
    setShareAllocator(delegate) {
        return new TransactionEnvelope(this.sdk.provider, [
            this.program.instruction.setShareAllocator({
                accounts: {
                    operator: this.key,
                    admin: this.sdk.provider.wallet.publicKey,
                    delegate,
                },
            }),
        ]);
    }
    get withDelegateAccounts() {
        return {
            operator: this.key,
            delegate: this.sdk.provider.wallet.publicKey,
            rewarder: this.data.rewarder,
            quarryMineProgram: this.sdk.programs.Mine.programId,
        };
    }
    delegateSetAnnualRewards(newAnnualRate) {
        return new TransactionEnvelope(this.sdk.provider, [
            this.program.instruction.delegateSetAnnualRewards(newAnnualRate, {
                accounts: {
                    withDelegate: this.withDelegateAccounts,
                },
            }),
        ]);
    }
    delegateSetFamine(newFamineTs, quarry) {
        return new TransactionEnvelope(this.sdk.provider, [
            this.program.instruction.delegateSetFamine(newFamineTs, {
                accounts: {
                    withDelegate: this.withDelegateAccounts,
                    quarry,
                },
            }),
        ]);
    }
    delegateCreateQuarry({ tokenMint, payer = this.sdk.provider.wallet.publicKey, }) {
        return __awaiter(this, void 0, void 0, function* () {
            const [quarry] = yield findQuarryAddress(this.data.rewarder, tokenMint, this.sdk.programs.Mine.programId);
            return {
                quarry,
                tx: new TransactionEnvelope(this.sdk.provider, [
                    this.program.instruction.delegateCreateQuarryV2({
                        accounts: {
                            withDelegate: this.withDelegateAccounts,
                            quarry,
                            tokenMint,
                            payer,
                            systemProgram: SystemProgram.programId,
                        },
                    }),
                ]),
            };
        });
    }
    delegateSetRewardsShare({ share, quarry, }) {
        return new TransactionEnvelope(this.sdk.provider, [
            this.program.instruction.delegateSetRewardsShare(new u64(share), {
                accounts: {
                    withDelegate: this.withDelegateAccounts,
                    quarry,
                },
            }),
        ]);
    }
}
