// @mercurial-finance/vault-sdk/dist/cjs/src/vault/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LOCKED_PROFIT_DEGRADATION_DENOMINATOR = exports.VAULT_STRATEGY_ADDRESS = exports.KEEPER_URL = exports.StrategyProgram = exports.SEEDS = exports.VAULT_BASE_KEY = exports.USDT_MINT = exports.USDC_MINT = exports.SOL_MINT = exports.REWARDER = exports.AFFILIATE_PROGRAM_ID = exports.PROGRAM_ID = void 0;
const web3_js_1 = require("@solana/web3.js");
const anchor_1 = require("@project-serum/anchor");
exports.PROGRAM_ID = '24Uqj9JCLxUeoC3hGfh5W3s9FM9uCHDS2SG3LYwBpyTi';
exports.AFFILIATE_PROGRAM_ID = 'GacY9YuN16HNRTy7ZWwULPccwvfFSBeNLuAQP7y38Du3';
exports.REWARDER = 'GuHrjvzqDvLTB27ebd9iFKwceCxKvSswzTByDQUTsvdm';
// Mainnet addresses
exports.SOL_MINT = new web3_js_1.PublicKey('So11111111111111111111111111111111111111112');
exports.USDC_MINT = new web3_js_1.PublicKey('EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v');
exports.USDT_MINT = new web3_js_1.PublicKey('Es9vMFrzaCERmJfrF4H2FYD4KCoNkY11McCe8BenwNYB');
exports.VAULT_BASE_KEY = new web3_js_1.PublicKey('HWzXGcGHy4tcpYfaRDCyLNzXqBTv3E6BttpCH2vJxArv');
exports.SEEDS = Object.freeze({
    VAULT_PREFIX: 'vault',
    TOKEN_VAULT_PREFIX: 'token_vault',
    LP_MINT_PREFIX: 'lp_mint',
    COLLATERAL_VAULT_PREFIX: 'collateral_vault',
    OBLIGATION_PREFIX: 'obligation',
    OBLIGATION_OWNER_PREFIX: 'obligation_owner',
    STAKING_PREFIX: 'staking',
    MINER: 'Miner',
    QUARRY: 'Quarry',
    APRICOT_USER_INFO_SIGNER_PREFIX: 'apricot_user_info_signer',
    FRAKT: 'frakt',
    DEPOSIT: 'deposit',
    FRAKT_LENDING: 'nftlendingv2',
    CYPHER: 'cypher',
    MANGO_ACCOUNT: 'MangoAccount',
    MANGO: 'mango',
    PSYLEND: 'psylend',
    PSYLEND_OWNER: 'deposits',
});
exports.StrategyProgram = {
    testnet: {
        solend: new web3_js_1.PublicKey('ALend7Ketfx5bxh6ghsCDXAoDrhvEmsXT3cynB6aPLgx'),
        portFinance: new web3_js_1.PublicKey('pdQ2rQQU5zH2rDgZ7xH2azMBJegUzUyunJ5Jd637hC4'),
    },
    devnet: {
        solend: new web3_js_1.PublicKey('ALend7Ketfx5bxh6ghsCDXAoDrhvEmsXT3cynB6aPLgx'),
        portFinance: new web3_js_1.PublicKey('pdQ2rQQU5zH2rDgZ7xH2azMBJegUzUyunJ5Jd637hC4'),
    },
    'mainnet-beta': {
        solend: new web3_js_1.PublicKey('So1endDq2YkqhipRh3WViPa8hdiSpxWy6z3Z6tMCpAo'),
        portFinance: new web3_js_1.PublicKey('Port7uDYB3wk6GJAw4KT1WpTeMtSu9bTcChBHkX2LfR'),
    },
};
exports.KEEPER_URL = {
    testnet: 'https://staging-keeper.raccoons.dev',
    devnet: 'https://dev-keeper.raccoons.dev',
    'mainnet-beta': 'https://merv2-api.mercurial.finance',
};
exports.VAULT_STRATEGY_ADDRESS = '11111111111111111111111111111111';
exports.LOCKED_PROFIT_DEGRADATION_DENOMINATOR = new anchor_1.BN(1000000000000);
