// @jup-ag/dradex-sdk/dist/cjs/examples/market_order.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const anchor_1 = require("@project-serum/anchor");
const nodewallet_1 = __importDefault(require("@project-serum/anchor/dist/cjs/nodewallet"));
const web3_js_1 = require("@solana/web3.js");
const bs58_1 = __importDefault(require("bs58"));
const __1 = require("..");
// ENTER YOUR SECRET KEY HERE
const SECRET_KEY = process.env.SECRET_KEY || "<your secret key>";
const KEYPAIR = web3_js_1.Keypair.fromSecretKey(bs58_1.default.decode(SECRET_KEY));
// GMT-USDC
const MARKET_ID = new web3_js_1.PublicKey("4v57hiqDBBpmpBU3ta3vmrgE4d8AWdWHyykLzzcfukJp");
async function main() {
    const connection = new web3_js_1.Connection("https://api.mainnet-beta.solana.com", "confirmed");
    const provider = new anchor_1.Provider(connection, new nodewallet_1.default(KEYPAIR), {
        commitment: "confirmed",
    });
    const dex = new __1.DexProgram({ provider });
    const publicKey = provider.wallet.publicKey;
    // fetch market
    const market = await dex.getMarket(MARKET_ID);
    await dex.loadMarket(market);
    // ensure that the MarketUser & DexUser exists
    const dexUserAddress = await dex.getDexUserAddress(publicKey);
    const marketUserAddress = await dex.getMarketUserAddress(market.address, publicKey);
    const [dexUser, marketUser] = await connection.getMultipleAccountsInfo([dexUserAddress, marketUserAddress]);
    const ix = dex.createInstructionSet([]);
    if (!dexUser) {
        ix.add(await dex.createDexUser(publicKey));
    }
    if (!marketUser) {
        ix.add(await dex.createMarketUser(market, publicKey));
    }
    // send a market order to buy 0.01 GMT
    // associated token accounts for both tokens must be available
    ix.add(await dex.createOrder(market, {
        amount: new anchor_1.BN(10000000),
        clientOrderId: new anchor_1.BN(0),
        limitPrice: new anchor_1.BN(0),
        limitTotal: null,
        minAmountOut: new anchor_1.BN(0),
        orderType: 0,
        side: __1.Side.BID,
    }));
    const txId = await ix.exec({ skipPreflight: true });
    console.log("txId:", txId);
}
main();
