// @jup-ag/dradex-sdk/dist/cjs/examples/market_quote.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const spl_token_1 = require("@solana/spl-token");
const web3_js_1 = require("@solana/web3.js");
const jsbi_1 = __importDefault(require("jsbi"));
const __1 = require("..");
const MARKET_ID = new web3_js_1.PublicKey("CJept8TLyG9r2GMhttqR98zohVq274XCMiy5oxjTKKBt");
const USDC_MINT = new web3_js_1.PublicKey("EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v");
async function main() {
    const connection = new web3_js_1.Connection("https://api.mainnet-beta.solana.com", "confirmed");
    // fetch market
    const marketAccountInfo = await connection.getAccountInfo(MARKET_ID);
    const market = new __1.DexMarket(MARKET_ID, marketAccountInfo);
    // update data
    const accountKeys = market.getAccountsForUpdate();
    const accounts = await connection.getMultipleAccountsInfo(accountKeys);
    const accountMap = toAccountInfoMap(accountKeys, accounts);
    market.update(accountMap);
    // get quotes
    const quotes = [
        [
            "BUY 1 USD of SOL",
            market.getQuote({
                amount: jsbi_1.default.BigInt(1000000),
                sourceMint: USDC_MINT,
                destinationMint: spl_token_1.NATIVE_MINT,
                swapMode: __1.SwapMode.ExactIn,
            }),
        ],
        [
            "SELL 1 SOL",
            market.getQuote({
                amount: jsbi_1.default.BigInt(1000000000),
                sourceMint: spl_token_1.NATIVE_MINT,
                destinationMint: USDC_MINT,
                swapMode: __1.SwapMode.ExactIn,
            }),
        ],
        [
            "SELL 1 USD of SOL",
            market.getQuote({
                amount: jsbi_1.default.BigInt(1000000),
                sourceMint: spl_token_1.NATIVE_MINT,
                destinationMint: USDC_MINT,
                swapMode: __1.SwapMode.ExactOut,
            }),
        ],
        [
            "BUY 1 SOL",
            market.getQuote({
                amount: jsbi_1.default.BigInt(1000000000),
                sourceMint: USDC_MINT,
                destinationMint: spl_token_1.NATIVE_MINT,
                swapMode: __1.SwapMode.ExactOut,
            }),
        ],
    ];
    quotes.forEach(([message, quote]) => {
        console.log(message, JSON.stringify(quote, undefined, 2), "\n");
    });
}
function toAccountInfoMap(keys, accounts) {
    const accountMap = new Map();
    accounts.forEach((account, index) => account && accountMap.set(keys[index].toString(), account));
    return accountMap;
}
main();
