// @mercurial-finance/francium-sdk/dist/index.mjs
// src/constants/lend/pools.ts
import { NATIVE_MINT as NATIVE_MINT2 } from "@solana/spl-token";
import { PublicKey as PublicKey2 } from "@solana/web3.js";

// src/constants/tokens/index.ts
import { NATIVE_MINT } from "@solana/spl-token";
import { PublicKey } from "@solana/web3.js";
var TOKENS = {
  SOL: {
    symbol: "SOL",
    name: "solana",
    mintAddress: NATIVE_MINT,
    decimals: 9
  },
  WSOL: {
    symbol: "WSOL",
    name: "Wrapped Solana",
    mintAddress: new PublicKey("So11111111111111111111111111111111111111112"),
    decimals: 9
  },
  BTC: {
    symbol: "BTC",
    name: "bitcoin",
    mintAddress: new PublicKey("9n4nbM75f5Ui33ZbPYXn59EwSgE8CGsHtAeTH5YFeJ9E"),
    decimals: 6
  },
  ETH: {
    symbol: "ETH",
    name: "ethereum",
    mintAddress: new PublicKey("2FPyTwcZLUg1MDrwsyoP4D6s1tM7hAkHYRjkNb5w6Pxk"),
    decimals: 6
  },
  USDT: {
    symbol: "USDT",
    name: "USDT",
    mintAddress: new PublicKey("Es9vMFrzaCERmJfrF4H2FYD4KCoNkY11McCe8BenwNYB"),
    decimals: 6
  },
  WUSDT: {
    symbol: "WUSDT",
    name: "Wrapped USDT",
    mintAddress: new PublicKey("BQcdHdAQW1hczDbBi9hiegXAR7A98Q9jx3X3iBBBDiq4"),
    decimals: 6
  },
  USDC: {
    symbol: "USDC",
    name: "USDC",
    mintAddress: new PublicKey("EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v"),
    decimals: 6
  },
  WUSDC: {
    symbol: "WUSDC",
    name: "Wrapped USDC",
    mintAddress: new PublicKey("BXXkv6z8ykpG1yuvUDPgh732wzVHB69RnB9YgSYh3itW"),
    decimals: 6
  },
  YFI: {
    symbol: "YFI",
    name: "Wrapped YFI",
    mintAddress: new PublicKey("3JSf5tPeuscJGtaCp5giEiDhv51gQ4v3zWg8DGgyLfAB"),
    decimals: 6
  },
  LINK: {
    symbol: "LINK",
    name: "Wrapped Chainlink",
    mintAddress: new PublicKey("CWE8jPTUYhdCTZYWPTe1o5DFqfdjzWKc9WKz6rSjQUdG"),
    decimals: 6
  },
  ALEPH: {
    symbol: "ALEPH",
    name: "Wrapped ALEPH",
    mintAddress: new PublicKey("CsZ5LZkDS7h9TDKjrbL7VAwQZ9nsRu8vJLhRYfmGaN8K"),
    decimals: 6
  },
  SXP: {
    symbol: "SXP",
    name: "Wrapped SXP",
    mintAddress: new PublicKey("SF3oTvfWzEP3DTwGSvUXRrGTvr75pdZNnBLAH9bzMuX"),
    decimals: 6
  },
  HGET: {
    symbol: "HGET",
    name: "Wrapped HGET",
    mintAddress: new PublicKey("BtZQfWqDGbk9Wf2rXEiWyQBdBY1etnUUn6zEphvVS7yN"),
    decimals: 6
  },
  CREAM: {
    symbol: "CREAM",
    name: "Wrapped CREAM",
    mintAddress: new PublicKey("5Fu5UUgbjpUvdBveb3a1JTNirL8rXtiYeSMWvKjtUNQv"),
    decimals: 6
  },
  UBXT: {
    symbol: "UBXT",
    name: "Wrapped UBXT",
    mintAddress: new PublicKey("873KLxCbz7s9Kc4ZzgYRtNmhfkQrhfyWGZJBmyCbC3ei"),
    decimals: 6
  },
  HNT: {
    symbol: "HNT",
    name: "Wrapped HNT",
    mintAddress: new PublicKey("HqB7uswoVg4suaQiDP3wjxob1G5WdZ144zhdStwMCq7e"),
    decimals: 6
  },
  FRONT: {
    symbol: "FRONT",
    name: "Wrapped FRONT",
    mintAddress: new PublicKey("9S4t2NEAiJVMvPdRYKVrfJpBafPBLtvbvyS3DecojQHw"),
    decimals: 6
  },
  AKRO: {
    symbol: "AKRO",
    name: "Wrapped AKRO",
    mintAddress: new PublicKey("6WNVCuxCGJzNjmMZoKyhZJwvJ5tYpsLyAtagzYASqBoF"),
    decimals: 6
  },
  HXRO: {
    symbol: "HXRO",
    name: "Wrapped HXRO",
    mintAddress: new PublicKey("DJafV9qemGp7mLMEn5wrfqaFwxsbLgUsGVS16zKRk9kc"),
    decimals: 6
  },
  UNI: {
    symbol: "UNI",
    name: "Wrapped UNI",
    mintAddress: new PublicKey("DEhAasscXF4kEGxFgJ3bq4PpVGp5wyUxMRvn6TzGVHaw"),
    decimals: 6
  },
  SRM: {
    symbol: "SRM",
    name: "serum",
    mintAddress: new PublicKey("SRMuApVNdxXokk5GT7XD5cUUgXMBCoAz2LHeuAoKWRt"),
    decimals: 6
  },
  FTT: {
    symbol: "FTT",
    name: "Wrapped FTT",
    mintAddress: new PublicKey("AGFEad2et2ZJif9jaGpdMixQqvW5i81aBdvKe7PHNfz3"),
    decimals: 6
  },
  TOMO: {
    symbol: "TOMO",
    name: "Wrapped TOMO",
    mintAddress: new PublicKey("GXMvfY2jpQctDqZ9RoU3oWPhufKiCcFEfchvYumtX7jd"),
    decimals: 6
  },
  KARMA: {
    symbol: "KARMA",
    name: "Wrapped KARMA",
    mintAddress: new PublicKey("EcqExpGNFBve2i1cMJUTR4bPXj4ZoqmDD2rTkeCcaTFX"),
    decimals: 4
  },
  LUA: {
    symbol: "LUA",
    name: "Wrapped LUA",
    mintAddress: new PublicKey("EqWCKXfs3x47uVosDpTRgFniThL9Y8iCztJaapxbEaVX"),
    decimals: 6
  },
  MATH: {
    symbol: "MATH",
    name: "Wrapped MATH",
    mintAddress: new PublicKey("GeDS162t9yGJuLEHPWXXGrb1zwkzinCgRwnT8vHYjKza"),
    decimals: 6
  },
  KEEP: {
    symbol: "KEEP",
    name: "Wrapped KEEP",
    mintAddress: new PublicKey("GUohe4DJUA5FKPWo3joiPgsB7yzer7LpDmt1Vhzy3Zht"),
    decimals: 6
  },
  SWAG: {
    symbol: "SWAG",
    name: "Wrapped SWAG",
    mintAddress: new PublicKey("9F9fNTT6qwjsu4X4yWYKZpsbw5qT7o6yR2i57JF2jagy"),
    decimals: 6
  },
  FIDA: {
    symbol: "FIDA",
    name: "Bonfida",
    mintAddress: new PublicKey("EchesyfXePKdLtoiZSL8pBe8Myagyy8ZRqsACNCFGnvp"),
    decimals: 6
  },
  KIN: {
    symbol: "KIN",
    name: "KIN",
    mintAddress: new PublicKey("kinXdEcpDQeHPEuQnqmUgtYykqKGVFq6CeVX5iAHJq6"),
    decimals: 5
  },
  MAPS: {
    symbol: "MAPS",
    name: "MAPS",
    mintAddress: new PublicKey("MAPS41MDahZ9QdKXhVa4dWB9RuyfV4XqhyAZ8XcYepb"),
    decimals: 6
  },
  OXY: {
    symbol: "OXY",
    name: "OXY",
    mintAddress: new PublicKey("z3dn17yLaGMKffVogeFHQ9zWVcXgqgf3PQnDsNs2g6M"),
    decimals: 6
  },
  RAY: {
    symbol: "RAY",
    name: "raydium",
    mintAddress: new PublicKey("4k3Dyjzvzp8eMZWUXbBCjEvwSkkk59S5iCNLY3QrkX6R"),
    decimals: 6
  },
  COPE: {
    symbol: "COPE",
    name: "cope",
    mintAddress: new PublicKey("8HGyAAB1yoM1ttS7pXjHMa3dukTFGQggnFFH3hJZgzQh"),
    decimals: 6
  },
  STEP: {
    symbol: "STEP",
    name: "step-finance",
    mintAddress: new PublicKey("StepAscQoEioFxxWGnh2sLBDFp9d8rvKz2Yp39iDpyT"),
    decimals: 9
  },
  MEDIA: {
    symbol: "MEDIA",
    name: "media-network",
    mintAddress: new PublicKey("ETAtLmCmsoiEEKfNrHKJ2kYy3MoABhU6NQvpSfij5tDs"),
    decimals: 6
  },
  ROPE: {
    symbol: "ROPE",
    name: "ROPE",
    mintAddress: new PublicKey("8PMHT4swUMtBzgHnh5U564N5sjPSiUz2cjEQzFnnP1Fo"),
    decimals: 9
  },
  MER: {
    symbol: "MER",
    name: "Mercurial",
    mintAddress: new PublicKey("MERt85fc5boKw3BW1eYdxonEuJNvXbiMbs6hvheau5K"),
    decimals: 6
  },
  LIKE: {
    symbol: "LIKE",
    name: "only1",
    mintAddress: new PublicKey("3bRTivrVsitbmCTGtqwp7hxXPsybkjn4XLNtPsHqa3zR"),
    decimals: 9
  },
  ATLAS: {
    symbol: "ATLAS",
    name: "star-atlas",
    decimals: 8,
    mintAddress: new PublicKey("ATLASXmbPQxBUYbxPsV97usA3fPQYEqzQBUHgiFCUsXx")
  },
  POLIS: {
    symbol: "POLIS",
    name: "star-atlas-dao",
    decimals: 8,
    mintAddress: new PublicKey("poLisWXnNRwC6oBu1vHiuKQzFjGL4XDSu4g9qjz9qVk")
  },
  ORCA: {
    symbol: "ORCA",
    name: "orca",
    decimals: 6,
    mintAddress: new PublicKey("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE")
  },
  mSOL: {
    symbol: "mSOL",
    name: "msol",
    decimals: 9,
    mintAddress: new PublicKey("mSoLzYCxHdYgdzU16g5QSh3i5K3z3KZK7ytfqcJm7So")
  },
  MNDE: {
    symbol: "MNDE",
    name: "marinade",
    decimals: 9,
    mintAddress: new PublicKey("MNDEFzGvMt87ueuHvVU9VcTqsAP5b3fTGPsHuuPA5ey")
  },
  NINJA: {
    symbol: "NINJA",
    name: "ninja-protocol",
    decimals: 6,
    mintAddress: new PublicKey("FgX1WD9WzMU3yLwXaFSarPfkgzjLb2DZCqmkx9ExpuvJ")
  },
  whETH: {
    symbol: "whETH",
    name: "ethereum",
    mintAddress: new PublicKey("7vfCXTUXx5WJV5JADk17DUJ4ksgau7utNKj4b963voxs"),
    decimals: 8
  },
  SBR: {
    symbol: "SBR",
    name: "saber",
    mintAddress: new PublicKey("Saber2gLauYim4Mvftnrasomsv6NvAuncvMEZwcLpD1"),
    decimals: 6
  },
  LIQ: {
    symbol: "LIQ",
    name: "liq-protocol",
    mintAddress: new PublicKey("4wjPQJ6PrkC4dHhYghwJzGBVP78DkBzA2U3kHoFNBuhj"),
    decimals: 6
  },
  PORT: {
    symbol: "PORT",
    name: "port-finance",
    mintAddress: new PublicKey("PoRTjZMPXb9T7dyU7tpLEZRQj7e6ssfAE62j2oQuc6y"),
    decimals: 6
  },
  SAMO: {
    symbol: "SAMO",
    name: "samoyedcoin",
    mintAddress: new PublicKey("7xKXtg2CW87d97TXJSDpbD5jBkheTqA83TZRuJosgAsU"),
    decimals: 9
  },
  ABR: {
    symbol: "ABR",
    name: "allbridge",
    mintAddress: new PublicKey("a11bdAAuV8iB2fu7X6AxAvDTo1QZ8FXB3kk5eecdasp"),
    decimals: 9
  },
  weSUSHI: {
    symbol: "weSUSHI",
    name: "allbridge",
    mintAddress: new PublicKey("ChVzxWRmrTeSgwd3Ui3UumcN8KX7VK3WaD4KGeSKpypj"),
    decimals: 8
  },
  weUNI: {
    symbol: "weUNI",
    name: "allbridge",
    mintAddress: new PublicKey("8FU95xFJhUUkyyCLU13HSzDLs7oC4QZdXQHL6SCeab36"),
    decimals: 8
  },
  STARS: {
    symbol: "STARS",
    name: "starlaunch",
    mintAddress: new PublicKey("HCgybxq5Upy8Mccihrp7EsmwwFqYZtrHrsmsKwtGXLgW"),
    decimals: 6
  },
  weDYDX: {
    symbol: "weDYDX",
    name: "dydx",
    mintAddress: new PublicKey("4Hx6Bj56eGyw8EJrrheM6LBQAvVYRikYCWsALeTrwyRU"),
    decimals: 8
  },
  weAXS: {
    symbol: "weAXS",
    name: "axie-infinity",
    mintAddress: new PublicKey("HysWcbHiYY9888pHbaqhwLYZQeZrcQMXKQWRqS7zcPK5"),
    decimals: 8
  },
  weSHIB: {
    symbol: "weSHIB",
    name: "shiba-inu",
    mintAddress: new PublicKey("CiKu4eHsVrc1eueVQeHn7qhXTcVu95gSQmBpX4utjL9z"),
    decimals: 8
  },
  SLND: {
    symbol: "SLND",
    name: "solend",
    mintAddress: new PublicKey("SLNDpmoWTVADgEdndyvWzroNL7zSi1dF9PC3xHGtPwp"),
    decimals: 6
  },
  weMANA: {
    symbol: "weMANA",
    name: "decentraland",
    mintAddress: new PublicKey("7dgHoN8wBZCc5wbnQ2C47TDnBMAxG4Q5L3KjP67z8kNi"),
    decimals: 8
  },
  weSAND: {
    symbol: "weSAND",
    name: "the-sandbox",
    mintAddress: new PublicKey("49c7WuCZkQgc3M4qH8WuEUNXfgwupZf1xqWkDQ7gjRGt"),
    decimals: 8
  },
  CAVE: {
    symbol: "CAVE",
    name: "cave",
    mintAddress: new PublicKey("4SZjjNABoqhbd4hnapbvoEPEqT8mnNkfbEoAwALf1V8t"),
    decimals: 6
  },
  GENE: {
    symbol: "GENE",
    name: "genopets",
    mintAddress: new PublicKey("GENEtH5amGSi8kHAtQoezp1XEXwZJ8vcuePYnXdKrMYz"),
    decimals: 9
  },
  SONAR: {
    symbol: "SONAR",
    name: "sonarwatch",
    mintAddress: new PublicKey("sonarX4VtVkQemriJeLm6CKeW3GDMyiBnnAEMw1MRAE"),
    decimals: 9
  },
  DFL: {
    symbol: "DFL",
    name: "defi-land",
    mintAddress: new PublicKey("DFL1zNkaGPWm1BqAVqRjCZvHmwTFrEaJtbzJWgseoNJh"),
    decimals: 9
  },
  wbWBNB: {
    symbol: "wbWBNB",
    name: "binancecoin",
    mintAddress: new PublicKey("9gP2kCy3wA1ctvYWQk75guqXuHfrEomqydHLtcTCqiLa"),
    decimals: 8
  },
  REAL: {
    symbol: "REAL",
    name: "realy-metaverse",
    mintAddress: new PublicKey("AD27ov5fVU2XzwsbvnFvb1JpCBaCB5dRXrczV9CqSVGb"),
    decimals: 9
  },
  MBS: {
    symbol: "MBS",
    name: "monkeyball",
    mintAddress: new PublicKey("Fm9rHUTF5v3hwMLbStjZXqNBBoZyGriQaFM6sTFz3K8A"),
    decimals: 6
  },
  stSOL: {
    symbol: "stSOL",
    name: "lido-staked-sol",
    mintAddress: new PublicKey("7dHbWXmci3dT8UFYWYZweBLXgycu7Y3iL6trKn1Y7ARj"),
    decimals: 9
  },
  wLDO: {
    symbol: "wLDO",
    name: "lido-dao",
    mintAddress: new PublicKey("HZRCwxP2Vq9PCpPXooayhJ2bxTpo5xfpQrwB1svh332p"),
    decimals: 8
  },
  PRISM: {
    symbol: "PRISM",
    name: "prism",
    mintAddress: new PublicKey("PRSMNsEPqhGVCH1TtWiJqPjJyh2cKrLostPZTNy1o5x"),
    decimals: 6
  },
  SHDW: {
    symbol: "SHDW",
    name: "genesysgo-shadow",
    mintAddress: new PublicKey("SHDWyBxihqiCj6YekG2GUr7wqKLeLAMK1gHZck9pL6y"),
    decimals: 9
  },
  BASIS: {
    symbol: "BASIS",
    name: "basis-markets",
    mintAddress: new PublicKey("Basis9oJw9j8cw53oMV7iqsgo6ihi9ALw4QR31rcjUJa"),
    decimals: 6
  },
  GRAPE: {
    symbol: "GRAPE",
    name: "grape-2",
    mintAddress: new PublicKey("8upjSpvjcdpuzhfR1zriwg5NXkwDruejqNE9WNbPRtyA"),
    decimals: 6
  },
  wUST: {
    symbol: "wUST",
    name: "wUST",
    mintAddress: new PublicKey("9vMJfxuKxXBoEa7rM12mYLMwTacLMLDJqHozw96WQL8i"),
    decimals: 6
  },
  wLUNA: {
    symbol: "wLUNA",
    name: "terra-luna",
    mintAddress: new PublicKey("F6v4wfAdJB8D8p77bMXZgYt8TDKsYxLYxH5AFhUkYx9W"),
    decimals: 6
  },
  SLC: {
    symbol: "SLC",
    name: "solice",
    mintAddress: new PublicKey("METAmTMXwdb8gYzyCPfXXFmZZw4rUsXX58PNsDg7zjL"),
    decimals: 6
  },
  MEAN: {
    symbol: "MEAN",
    name: "meanfi",
    mintAddress: new PublicKey("MEANeD3XDdUmNMsRGjASkSWdC8prLYsoRJ61pPeHctD"),
    decimals: 6
  },
  ZBC: {
    symbol: "ZBC",
    name: "zebec-protocol",
    mintAddress: new PublicKey("zebeczgi5fSEtbpfQKVZKCJ3WgYXxjkMUkNNx7fLKAF"),
    decimals: 9
  },
  ATS: {
    symbol: "ATS",
    name: "atlas-dex",
    mintAddress: new PublicKey("HJbNXx2YMRxgfUJ6K4qeWtjatMK5KYQT1QnsCdDWywNv"),
    decimals: 9
  },
  GST: {
    symbol: "GST",
    name: "green-satoshi-token",
    mintAddress: new PublicKey("AFbX8oGjGpmVFywbVouvhQSRmiW2aR1mohfahi4Y2AdB"),
    decimals: 9
  },
  GMT: {
    symbol: "GMT",
    name: "stepn",
    mintAddress: new PublicKey("7i5KKsX2weiTkry7jA4ZwSuXGhs5eJBEjY8vVxR4pfRx"),
    decimals: 9
  },
  prANA: {
    symbol: "prANA",
    name: "nirvana-ana",
    mintAddress: new PublicKey("PRAxfbouRoJ9yZqhyejEAH6RvjJ86Y82vfiZTBSM3xG"),
    decimals: 6
  },
  RATIO: {
    symbol: "RATIO",
    name: "ratio-finance",
    mintAddress: new PublicKey("ratioMVg27rSZbSvBopUvsdrGUzeALUfFma61mpxc8J"),
    decimals: 6
  },
  TTT: {
    symbol: "TTT",
    name: "tabtrader",
    mintAddress: new PublicKey("FNFKRV3V8DtA3gVJN6UshMiLGYA8izxFwkNWmJbFjmRj"),
    decimals: 6
  },
  RUN: {
    symbol: "RUN",
    name: "run",
    mintAddress: new PublicKey("6F9XriABHfWhit6zmMUYAQBSy6XK5VF1cHXuW5LDpRtC"),
    decimals: 9
  },
  SLCL: {
    symbol: "SLCL",
    name: "solcial",
    mintAddress: new PublicKey("SLCLww7nc1PD2gQPQdGayHviVVcpMthnqUz2iWKhNQV"),
    decimals: 9
  }
};

// src/constants/lend/pools.ts
var lendProgramId = new PublicKey2(
  "FC81tbGt6JWRXidaWYFXxGnTk4VgobhJHATvTRVMqgWj"
);
var lendingPools = {
  USDC: {
    lookupTableAddress: new PublicKey2(
      "8zkUc4NeXDoFp8UBDQ3sPVKbBAHNHdvNbeRm1tupEHS9"
    ),
    programId: lendProgramId,
    tokenMint: new PublicKey2("EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v"),
    marketInfoAccount: new PublicKey2(
      "4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"
    ),
    marketOwner: new PublicKey2("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
    lendingPoolInfoAccount: new PublicKey2(
      "Hx6LbkMHe69DYawhPyVNs8Apa6tyfogfzQV6a7XkwBUU"
    ),
    lendingMarketAuthority: new PublicKey2(
      "sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"
    ),
    lendingPoolTknAccount: new PublicKey2(
      "CFp9kt8z3Epb1QSiEp3xA44KbSwuJxhFR3wQoerFqYS9"
    ),
    lendingPoolFeeAccount: new PublicKey2(
      "CugsLyJk1Jghc7LkgvnFuUwshJhz1FE9mpkF7Z4acAMU"
    ),
    lendingPoolShareMint: new PublicKey2(
      "62fDf5daUJ9jBz8Xtj6Bmw1bh1DvHn8AG4L9hMmxCzpu"
    ),
    lendingPoolShareAccount: new PublicKey2(
      "Gyc1V1xbA9NjzuURE662ATw6W4AdhwvsL26yUnSGhbkz"
    ),
    lendingPoolCreditMint: new PublicKey2(
      "CZ2s85dnuAVyGbRWBNauHZwj9oTV2i9xTvFwc1Cedr7E"
    ),
    lendingPoolCreditAccount: new PublicKey2(
      "Bd71jEvypKFs8N5gByyhD2149tSqLy4PVAzFFShCTJTB"
    )
  },
  PAI: {
    lookupTableAddress: new PublicKey2(
      "2E8tZ5SLctXECt74ooLxDf4Wz5dwngpoKwc6z7Mn987S"
    ),
    programId: lendProgramId,
    tokenMint: new PublicKey2("Ea5SjE2Y6yvCeW5dYTn7PYMuW5ikXkvbGdcmSnXeaLjS"),
    marketInfoAccount: new PublicKey2(
      "4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"
    ),
    marketOwner: new PublicKey2("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
    lendingPoolInfoAccount: new PublicKey2(
      "EzofB5BK23PHDfEAThk5oJANb9FWTXzezrafiTDzcBBA"
    ),
    lendingMarketAuthority: new PublicKey2(
      "sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"
    ),
    lendingPoolTknAccount: new PublicKey2(
      "4Kvd1ULAy45k2EQt1pdePa82UBz9tW6N1rLJWb3EAJgQ"
    ),
    lendingPoolFeeAccount: new PublicKey2(
      "sKmPL3zkAf7777bhXnMaSfTU1jg1keNHNPz38wBTNd5"
    ),
    lendingPoolShareMint: new PublicKey2(
      "HDvD8a4VWbkHNG7hb4CBumNhn41DyKL51qVYBNH73o23"
    ),
    lendingPoolShareAccount: new PublicKey2(
      "9tk9XwzJt1J2DoNcBtBmGMsnbnwvMUAoJF1pjdcfJe2E"
    ),
    lendingPoolCreditMint: new PublicKey2(
      "HkMze549cFMdLMRDsJyfH8n6EE5jjTuVZr9re4WRYuk9"
    ),
    lendingPoolCreditAccount: new PublicKey2(
      "7h8YziEXSZgyzqDb5c6PP3uUztQR7hQHwqyWWe9BTqY3"
    )
  },
  RAY: {
    lookupTableAddress: new PublicKey2(
      "Fe8sni23YSFGqZTLpTia6y6fUmbyzao22wJy1hBqGSx"
    ),
    programId: lendProgramId,
    tokenMint: new PublicKey2("4k3Dyjzvzp8eMZWUXbBCjEvwSkkk59S5iCNLY3QrkX6R"),
    marketInfoAccount: new PublicKey2(
      "4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"
    ),
    marketOwner: new PublicKey2("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
    lendingPoolInfoAccount: new PublicKey2(
      "ENZ7vdrvNGdAfVa8DoGpT2GrfoSFUtWKRoesf2dvqpuq"
    ),
    lendingMarketAuthority: new PublicKey2(
      "sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"
    ),
    lendingPoolTknAccount: new PublicKey2(
      "BUK8XbZeEqVULYRxGp4Va2R1BRqrYzXDrCVnWt6B7TZf"
    ),
    lendingPoolFeeAccount: new PublicKey2(
      "C2HJD7r6NYCCVoXPPExxYEQ73LZJ3qmb5H13GnwMZfwa"
    ),
    lendingPoolShareMint: new PublicKey2(
      "5uZERkJVyhCABEdocEssPZyMXYa8GJpLhngi95yXr9jc"
    ),
    lendingPoolShareAccount: new PublicKey2(
      "GrVpvHCRkTHESGwyaj4gtNLn95NAdkBSnZbKusoFumup"
    ),
    lendingPoolCreditMint: new PublicKey2(
      "GZqhoJhN1bz5Frq9GCaUSJ2nimnizXzHwsH12Qgh4YmR"
    ),
    lendingPoolCreditAccount: new PublicKey2(
      "6EZzqDqfTfdoFF4KD1gmki3btEHfULK4SjNcuJMTEDUt"
    )
  },
  USDT: {
    lookupTableAddress: new PublicKey2(
      "8T9RNbWAWuPgKXgV5EaQV8kvo9fbGjmLmt4XzwFimn1c"
    ),
    programId: lendProgramId,
    tokenMint: new PublicKey2("Es9vMFrzaCERmJfrF4H2FYD4KCoNkY11McCe8BenwNYB"),
    marketInfoAccount: new PublicKey2(
      "4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"
    ),
    marketOwner: new PublicKey2("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
    lendingPoolInfoAccount: new PublicKey2(
      "EyqWUeme2B1dhuwJvJVi2coYPcGWmWrakcpsSFeikrhb"
    ),
    lendingMarketAuthority: new PublicKey2(
      "sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"
    ),
    lendingPoolTknAccount: new PublicKey2(
      "Acx3Ky9pk8CyYftA9HYF31EbVhBDi21EqJsFtMW2XxXW"
    ),
    lendingPoolFeeAccount: new PublicKey2(
      "7gAN5Y4hGAG2A9vyr9dURGosjmo4Rwiaqz4iKZGNenE1"
    ),
    lendingPoolShareMint: new PublicKey2(
      "8TtsTZQ6mU4YSW5jw4sDjSDEG1CW8QLpK7C9g5TVgBvn"
    ),
    lendingPoolShareAccount: new PublicKey2(
      "EfLgyXRGFi9MUCMgrsFBuAJ7pMPXBceKhD6x4PgdvLLc"
    ),
    lendingPoolCreditMint: new PublicKey2(
      "CoheXVD8cAdVaKznfay22dBpfML5Fbz5g4j67syKgRg4"
    ),
    lendingPoolCreditAccount: new PublicKey2(
      "EAQxDXsA6RXxWtA1PwZNnTnj2MapHBtMzvUmfuGPJCPX"
    )
  },
  SRM: {
    lookupTableAddress: new PublicKey2(
      "E7Rf4v3DroXQ4Q2Aon5FsqtDvduNuXEXqCcoKoD8UYoj"
    ),
    programId: lendProgramId,
    tokenMint: new PublicKey2("SRMuApVNdxXokk5GT7XD5cUUgXMBCoAz2LHeuAoKWRt"),
    marketInfoAccount: new PublicKey2(
      "4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"
    ),
    marketOwner: new PublicKey2("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
    lendingPoolInfoAccount: new PublicKey2(
      "B1zB1EuTjnFPLdwySeBYhzeAf3h9buWLbDoG7AHcUTMF"
    ),
    lendingMarketAuthority: new PublicKey2(
      "sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"
    ),
    lendingPoolTknAccount: new PublicKey2(
      "FLQtBThLEVvhXdKqq2CREL8sFt8jAFS8szm4HaMYqmJk"
    ),
    lendingPoolFeeAccount: new PublicKey2(
      "FdSBYwMEfy21H3k64cvYdir3mxzmnZotiYUsC5iPPoxM"
    ),
    lendingPoolShareMint: new PublicKey2(
      "6CZhCFwA7hbqDEoVUQdFpHaCGQU7fSSwhaJWdG4DmsQ6"
    ),
    lendingPoolShareAccount: new PublicKey2(
      "C3EjtH3hVuLrU3j1y5ArMFRbAhxFf5hXNfk3b9SU91qN"
    ),
    lendingPoolCreditMint: new PublicKey2(
      "DqKuxgMxLPDCXtoQzLG1p85S2i3tP3EaeUVDAw7P8Nt4"
    ),
    lendingPoolCreditAccount: new PublicKey2(
      "Ednqn9Zo5HpDX2DRmCWu6x46ZXnHNmTmrq9G8hrTuBnz"
    )
  },
  ETH: {
    lookupTableAddress: new PublicKey2(
      "5M7vfgfzqT4F5QJyb7w9YqGPrVXBLhQa7sqVRfmd5NE7"
    ),
    programId: lendProgramId,
    tokenMint: new PublicKey2("2FPyTwcZLUg1MDrwsyoP4D6s1tM7hAkHYRjkNb5w6Pxk"),
    marketInfoAccount: new PublicKey2(
      "4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"
    ),
    marketOwner: new PublicKey2("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
    lendingPoolInfoAccount: new PublicKey2(
      "CKMQxUz1nkn3NS5B9AUD1uyWNL8iN2piG9LVt1RvWXzj"
    ),
    lendingMarketAuthority: new PublicKey2(
      "sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"
    ),
    lendingPoolTknAccount: new PublicKey2(
      "9MH38iiDX4Pk37U6TXLqz2783RspNhwBTYwBNHp8WUzP"
    ),
    lendingPoolFeeAccount: new PublicKey2(
      "8UPe7Fcm2f1QEFQh2YNr1jg2vgQmj4CXhLYEWgStHd8B"
    ),
    lendingPoolShareMint: new PublicKey2(
      "B8QXcUv5FFwyHH5V88g6PhYBc8fQvwSpza4C9PsiRpQD"
    ),
    lendingPoolShareAccount: new PublicKey2(
      "C5X2Q2K2jQtwpuqHKnLVJ1ZsvL9BMRwddMgqaQ5UGNkC"
    ),
    lendingPoolCreditMint: new PublicKey2(
      "BKKbCDggM83SPLFd1jRPu1ZsCv9nXkwZVne11N9FEs9w"
    ),
    lendingPoolCreditAccount: new PublicKey2(
      "9A3KAmmv1VyqNqVGBM6T9b1dp9Ax9xxdeXEgedo8U7Gh"
    )
  },
  SOL: {
    lookupTableAddress: new PublicKey2(
      "4DoNrJQqMB2kG31bzKxVdgKWqB4rrCNt3bVbAzUd4Tmg"
    ),
    programId: lendProgramId,
    tokenMint: NATIVE_MINT2,
    marketInfoAccount: new PublicKey2(
      "4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"
    ),
    marketOwner: new PublicKey2("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
    lendingPoolInfoAccount: new PublicKey2(
      "EnZC8MB6QLTxwN1LNqCXYC7XMpvXqitnYQPf5y5AcQRn"
    ),
    lendingMarketAuthority: new PublicKey2(
      "sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"
    ),
    lendingPoolTknAccount: new PublicKey2(
      "GiJwrxQW5NQWWAva4rnDMXwHxfrhyUzwMX5to1pmCCgF"
    ),
    lendingPoolFeeAccount: new PublicKey2(
      "2BDZqoVKAX3PJKeZJytjk7oEpbUu3jvdgdi6tA8VusnY"
    ),
    lendingPoolShareMint: new PublicKey2(
      "92Zst3rEoJsyZUW2yNc2811GtkBwFrm3tqgxMTtFWuSB"
    ),
    lendingPoolShareAccount: new PublicKey2(
      "Hyi1jMgc9MMEZyaG9ziL2PvT7zbeRNKhb5K1kj9bxz4x"
    ),
    lendingPoolCreditMint: new PublicKey2(
      "7x61NrCzeekHwgW5Nk4afMg2sxjMK6RyMkx8tCzzX4qG"
    ),
    lendingPoolCreditAccount: new PublicKey2(
      "WuPCgCuyLSXatdjyWqkGjCPSQFgKLwcfsNR9fDq3hVL"
    )
  },
  ORCA: {
    lookupTableAddress: new PublicKey2(
      "2iiSgXsQv5Ej6m68KkP7uU1jCZwy2Py1RhwLkG91PBt4"
    ),
    programId: lendProgramId,
    tokenMint: new PublicKey2("orcaEKTdK7LKz57vaAYr9QeNsVEPfiu6QeMU1kektZE"),
    marketInfoAccount: new PublicKey2(
      "4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"
    ),
    marketOwner: new PublicKey2("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
    lendingPoolInfoAccount: new PublicKey2(
      "9bo5JunUhp4XD7TDeeS9ARvWTYGBYtuQKUHGUb7RRvkf"
    ),
    lendingMarketAuthority: new PublicKey2(
      "sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"
    ),
    lendingPoolTknAccount: new PublicKey2(
      "F3EhkuRuZaYNY2bLTosjv6V7QXiZ8db9nXTZeLnmHC7V"
    ),
    lendingPoolFeeAccount: new PublicKey2(
      "F3EhkuRuZaYNY2bLTosjv6V7QXiZ8db9nXTZeLnmHC7V"
    ),
    lendingPoolShareMint: new PublicKey2(
      "DKoKBD7YheCGZjWp5CaHDPtheAcUveRSMvWkaivbNzWh"
    ),
    lendingPoolShareAccount: new PublicKey2(
      "AMUE9EgiEiTP1YVBicwTVGtumBypR5thjTgbsXZM2PPG"
    ),
    lendingPoolCreditMint: new PublicKey2(
      "CJNd1LZZxZr243dpNN9DtEavXn65kstrg251MPb8vmwW"
    ),
    lendingPoolCreditAccount: new PublicKey2(
      "74m1zyzV8x9d8kZLEaVaNWRsnx8pH47tfX8cquot11ef"
    )
  },
  mSOL: {
    lookupTableAddress: new PublicKey2(
      "A9L7C5QMNG5yBjntCxjo4tkxmu2AHiNP5ev9uBtLUMfE"
    ),
    programId: lendProgramId,
    tokenMint: new PublicKey2("mSoLzYCxHdYgdzU16g5QSh3i5K3z3KZK7ytfqcJm7So"),
    marketInfoAccount: new PublicKey2(
      "4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"
    ),
    marketOwner: new PublicKey2("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
    lendingMarketAuthority: new PublicKey2(
      "sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"
    ),
    lendingPoolInfoAccount: new PublicKey2(
      "492Hr5QDQyrsLcyXqMZ7A5osjmqwtn4cdKNQLDBYBzvy"
    ),
    lendingPoolTknAccount: new PublicKey2(
      "ADW9ZJuRQ9xbzTtWCwX1Th24Vxq7GWHWwYtctrbyjSr1"
    ),
    lendingPoolFeeAccount: new PublicKey2(
      "ADW9ZJuRQ9xbzTtWCwX1Th24Vxq7GWHWwYtctrbyjSr1"
    ),
    lendingPoolShareMint: new PublicKey2(
      "4mntRxKJWib1YTWNV4aLfszF6FeziBHdwnNAwGSfKKMX"
    ),
    lendingPoolShareAccount: new PublicKey2(
      "EbWWhmDLKXDdq2guubVJCdSztuUaN6XhtwVMy9TMMwtV"
    ),
    lendingPoolCreditMint: new PublicKey2(
      "Fdu3QfaQ9VL4GF85NrkpzhKRAJECv9cx7gchHca5MsRt"
    ),
    lendingPoolCreditAccount: new PublicKey2(
      "HdTwUP1U3WQ1EnYPP8c8mx8sqHhXYQ9zwKVsLGxWi7W1"
    )
  },
  BTC: {
    lookupTableAddress: new PublicKey2(
      "HUT8P5M4buU38F361SC8Zzp5VPAHfMbmL4aAzoyLdVp6"
    ),
    programId: lendProgramId,
    tokenMint: new PublicKey2("9n4nbM75f5Ui33ZbPYXn59EwSgE8CGsHtAeTH5YFeJ9E"),
    marketInfoAccount: new PublicKey2(
      "4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"
    ),
    marketOwner: new PublicKey2("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
    lendingMarketAuthority: new PublicKey2(
      "sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"
    ),
    lendingPoolInfoAccount: new PublicKey2(
      "DxAuEAxKYXsKMYG5Ma4TM8AsrAo7LQBSnbEGicU7i1ek"
    ),
    lendingPoolTknAccount: new PublicKey2(
      "2MN34BxuLVrjZpKYKHmW1c6ZGeQs4aWQRrEvmrYfZdtG"
    ),
    lendingPoolFeeAccount: new PublicKey2(
      "2MN34BxuLVrjZpKYKHmW1c6ZGeQs4aWQRrEvmrYfZdtG"
    ),
    lendingPoolShareMint: new PublicKey2(
      "2G9iwy9zfLaXB2bFiqSA7YbKEvtAEXVmdvGTF28jQVgg"
    ),
    lendingPoolShareAccount: new PublicKey2(
      "HGejymLhai1TDvRA4vBEbS5VRbKP1u74ZdzZcv8Y2o1M"
    ),
    lendingPoolCreditMint: new PublicKey2(
      "6WzqPd81AqsCwbXp33A6ro87axfD9je32hn3JE4KsnGi"
    ),
    lendingPoolCreditAccount: new PublicKey2(
      "E7HkZYeY8mqrtC45kusWiArqLNpGnV5Foioj6yEPHB1y"
    )
  },
  whETH: {
    lookupTableAddress: new PublicKey2(
      "9YyTJSjV9YK4GgW4oKETJQcEJyHTCCDU1TwVFZu41Cjs"
    ),
    programId: lendProgramId,
    tokenMint: new PublicKey2("7vfCXTUXx5WJV5JADk17DUJ4ksgau7utNKj4b963voxs"),
    marketInfoAccount: new PublicKey2(
      "4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"
    ),
    marketOwner: new PublicKey2("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
    lendingMarketAuthority: new PublicKey2(
      "sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"
    ),
    lendingPoolInfoAccount: new PublicKey2(
      "EEokU6r9bBdTz1umHjGhkjgfikARsJzJBXhYxnTmN4Xk"
    ),
    lendingPoolTknAccount: new PublicKey2(
      "9Mq6KEyW1nF24TcGrdLLfnK2pE5VdMxbY2SZtcP84R5X"
    ),
    lendingPoolFeeAccount: new PublicKey2(
      "9Mq6KEyW1nF24TcGrdLLfnK2pE5VdMxbY2SZtcP84R5X"
    ),
    lendingPoolShareMint: new PublicKey2(
      "Hk8b6i2C7PxFBPdd1TCrGssuZpCDEg4fh8yFuyHNJuJe"
    ),
    lendingPoolShareAccount: new PublicKey2(
      "FNG4A4GWF1EsQVEjNrQMRT9Dr3H28TycQabtnsQF1MiD"
    ),
    lendingPoolCreditMint: new PublicKey2(
      "7f7mU3tQvxD9t3f5ckhxcGSdwqoHHdWrJg43HXiUFWri"
    ),
    lendingPoolCreditAccount: new PublicKey2(
      "7LvrwstXyshwVn7BPDgjR8Cndersy9GzbbGhyHwQcX3W"
    )
  },
  weSUSHI: {
    lookupTableAddress: new PublicKey2(
      "5vnSYw9FpCWYf6NB1VxNhC9ouv7d1xLz9Cx17S2f46aW"
    ),
    programId: lendProgramId,
    tokenMint: TOKENS.weSUSHI.mintAddress,
    marketInfoAccount: new PublicKey2(
      "4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"
    ),
    marketOwner: new PublicKey2("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
    lendingMarketAuthority: new PublicKey2(
      "sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"
    ),
    lendingPoolInfoAccount: new PublicKey2(
      "2oTEVdMpSRsamFeNYzqn3wifsPHRomA8TShgbTnDtTgN"
    ),
    lendingPoolTknAccount: new PublicKey2(
      "4VAyWCoRbW5YXiV5Rz7pZy7miYBJLjfZYsykDuTnWyqj"
    ),
    lendingPoolFeeAccount: new PublicKey2(
      "4VAyWCoRbW5YXiV5Rz7pZy7miYBJLjfZYsykDuTnWyqj"
    ),
    lendingPoolShareMint: new PublicKey2(
      "EEhiV55jAt5JDpeH3GF4VGrStiPn5gCeWmqffyTp9B4E"
    ),
    lendingPoolShareAccount: new PublicKey2(
      "5SDoskkCcV8NBJcP97g9CS7BEGLqXT5q2F6Ve2W3VN71"
    ),
    lendingPoolCreditMint: new PublicKey2(
      "E2ocWnUUuBpUDLRSxB5VdUqr635DMPiWuV4PYDtiyGbi"
    ),
    lendingPoolCreditAccount: new PublicKey2(
      "L9idBQvg4SKzdtHPgZijpJGkDT4WVbhnA7KmTVPT477"
    )
  },
  weUNI: {
    lookupTableAddress: new PublicKey2(
      "Hb4ZHgHCyyzJWfksKyr8mRB7ydvH2SG76AC5pXkv6j9G"
    ),
    programId: lendProgramId,
    tokenMint: TOKENS.weUNI.mintAddress,
    marketInfoAccount: new PublicKey2(
      "4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"
    ),
    marketOwner: new PublicKey2("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
    lendingMarketAuthority: new PublicKey2(
      "sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"
    ),
    lendingPoolInfoAccount: new PublicKey2(
      "H4uTQ8CCFJYVQYZ8c2bH7hHGrrok6k4pWDhk5NQt9KR8"
    ),
    lendingPoolTknAccount: new PublicKey2(
      "E3fN9Wqesn2NbYGPjybFo5HTMpcyLPnWNHovQpQqqw8G"
    ),
    lendingPoolFeeAccount: new PublicKey2(
      "E3fN9Wqesn2NbYGPjybFo5HTMpcyLPnWNHovQpQqqw8G"
    ),
    lendingPoolShareMint: new PublicKey2(
      "CGj7e1g4ojW1RhSocF8AKWxHFaqCQjK8kBsbHqNB4BxP"
    ),
    lendingPoolShareAccount: new PublicKey2(
      "8BtR6Dzg4fLmLv4ZZdX1x22WYiPHvp8AtvRFgSMQwxdw"
    ),
    lendingPoolCreditMint: new PublicKey2(
      "74WmdEZG9rjvcAX2yczd5h54QeXnq62sVNLMz1DU2Abh"
    ),
    lendingPoolCreditAccount: new PublicKey2(
      "8kLAk9EiLV6qePNEtjnDLyWbRsySgMPSTihdWAEEv9H9"
    )
  },
  SAMO: {
    lookupTableAddress: new PublicKey2(
      "FUdXgKTfisX642gKDZReFtaRee8WdHiw4nrz4Y4y7JpP"
    ),
    programId: lendProgramId,
    tokenMint: TOKENS.SAMO.mintAddress,
    marketInfoAccount: new PublicKey2(
      "4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"
    ),
    marketOwner: new PublicKey2("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
    lendingMarketAuthority: new PublicKey2(
      "sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"
    ),
    lendingPoolInfoAccount: new PublicKey2(
      "HJLkovTpxof1z9tVMuX1pWp1ePDx1zgmnGYfFMxFRAd"
    ),
    lendingPoolTknAccount: new PublicKey2(
      "J198uwytk8CPejCyNM9TMPJcLPCky6pZwGwP3n1CUfto"
    ),
    lendingPoolFeeAccount: new PublicKey2(
      "J198uwytk8CPejCyNM9TMPJcLPCky6pZwGwP3n1CUfto"
    ),
    lendingPoolShareMint: new PublicKey2(
      "A9H3fAqkWmRnnFzXXzydZHzyLQdzK5o9dMejCL27tqq8"
    ),
    lendingPoolShareAccount: new PublicKey2(
      "HP3sKy5xqLkoNnbuHRCPuFaQUtYskNnockBPcYo5PZut"
    ),
    lendingPoolCreditMint: new PublicKey2(
      "E5ejPudazVxEdQUsS37g8yUn5HD22FBHsGEKWkGRDR3F"
    ),
    lendingPoolCreditAccount: new PublicKey2(
      "23tLfFHXZR1tAimzVyEAdN7HWJWiqvjepaX5qBHkgQvT"
    )
  },
  POLIS: {
    lookupTableAddress: new PublicKey2(
      "8GNhPXDU4UGascdfKGmu3ZBJsEbYoxe4w2BRixXH8ZNK"
    ),
    programId: lendProgramId,
    tokenMint: TOKENS.POLIS.mintAddress,
    marketInfoAccount: new PublicKey2(
      "4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"
    ),
    marketOwner: new PublicKey2("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
    lendingMarketAuthority: new PublicKey2(
      "sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"
    ),
    lendingPoolInfoAccount: new PublicKey2(
      "BWJkfNQnYt817V6sKxEJ3sFcivcXZgfKGCe8Z9f4LcVj"
    ),
    lendingPoolTknAccount: new PublicKey2(
      "7Ad6djoPLzigyBnnWjoLh2gQ6Tbw2s6kAdsaVUBq6L8N"
    ),
    lendingPoolFeeAccount: new PublicKey2(
      "7Ad6djoPLzigyBnnWjoLh2gQ6Tbw2s6kAdsaVUBq6L8N"
    ),
    lendingPoolShareMint: new PublicKey2(
      "FenVvq6s6S3McD1BCm76Ktz1EvRNCB4qYKGFU76fB7Fj"
    ),
    lendingPoolShareAccount: new PublicKey2(
      "GKTqMGVCgXJaDzjYfPdgMbVfnzCKDj6KqRpykauw19do"
    ),
    lendingPoolCreditMint: new PublicKey2(
      "EkKFNt7PBRdWy8EpmZAbAZdvXZpSKvdwMfo8eotN1PEr"
    ),
    lendingPoolCreditAccount: new PublicKey2(
      "FGKikDsnBPQ5m7jgbNNbHaR1gh1T8GYJ5sPYWoYdoE4p"
    )
  },
  ATLAS: {
    lookupTableAddress: new PublicKey2(
      "7jmphqSz19pSyYDAoNbBnAdpSPjCDMhuGowvZL4MVPya"
    ),
    programId: lendProgramId,
    tokenMint: TOKENS.ATLAS.mintAddress,
    marketInfoAccount: new PublicKey2(
      "4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"
    ),
    marketOwner: new PublicKey2("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
    lendingMarketAuthority: new PublicKey2(
      "sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"
    ),
    lendingPoolInfoAccount: new PublicKey2(
      "FnkmnUqazYmSm791rSTSXYbAzDzsuHJBgySgNZeCHGDJ"
    ),
    lendingPoolTknAccount: new PublicKey2(
      "E5ovYitzudRyUy7AS4U52eFgrq1rjr773rRGpYax1nmD"
    ),
    lendingPoolFeeAccount: new PublicKey2(
      "E5ovYitzudRyUy7AS4U52eFgrq1rjr773rRGpYax1nmD"
    ),
    lendingPoolShareMint: new PublicKey2(
      "4mygt5bFQrbXH9gNg75j1KVTrKGhvcYiQjir6FJ8afYH"
    ),
    lendingPoolShareAccount: new PublicKey2(
      "Eb4Ai2PXmRX3V9TX6awjNdBXZKGRyCgTmQu6Pni9NnTb"
    ),
    lendingPoolCreditMint: new PublicKey2(
      "3nAgm2XrSi3RNDWz4wCvUWwQW3QQE7s5i7MxNz8r8mGZ"
    ),
    lendingPoolCreditAccount: new PublicKey2(
      "3PwecBTLVt8zqKadSFDjeKH7Swzt1GxvU27urj66L89P"
    )
  },
  GENE: {
    lookupTableAddress: new PublicKey2(
      "HFKYALQcRFkM5S35fjXz3aCoyPDwKeGJtXzs5yMkjUFw"
    ),
    programId: lendProgramId,
    tokenMint: TOKENS.GENE.mintAddress,
    marketInfoAccount: new PublicKey2(
      "4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"
    ),
    marketOwner: new PublicKey2("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
    lendingMarketAuthority: new PublicKey2(
      "sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"
    ),
    lendingPoolInfoAccount: new PublicKey2(
      "FrgyPPEPNC25ihob4ZobP3eYjh8uBoxLiobdMDX2EsMc"
    ),
    lendingPoolTknAccount: new PublicKey2(
      "A4UMfr97y35AvCaakKRAE5UBsb4UuvRQ2JiT23dyzDFu"
    ),
    lendingPoolFeeAccount: new PublicKey2(
      "A4UMfr97y35AvCaakKRAE5UBsb4UuvRQ2JiT23dyzDFu"
    ),
    lendingPoolShareMint: new PublicKey2(
      "9z3eDHueAMdUtym9Q2ku3hi5YXHTYjpFLp1YEEnxUHPV"
    ),
    lendingPoolShareAccount: new PublicKey2(
      "58teG2GNEoeYQUBa3VxQe5ZH5SPCZ6nUsrCuN5akbaQS"
    ),
    lendingPoolCreditMint: new PublicKey2(
      "2KR5Q6zCik6kFyiWPMMVCKC5HDAP6joGNhKhbEv6nFdY"
    ),
    lendingPoolCreditAccount: new PublicKey2(
      "7LBnkdaCZsM82t7G5MYKBoYAZetcnsqfU1cABxnPoquw"
    )
  },
  SONAR: {
    lookupTableAddress: new PublicKey2(
      "CUcpP2uUw77Ftvhdppy2SpcRsDQ9kgW4qeG6iWNKeiC1"
    ),
    programId: lendProgramId,
    tokenMint: TOKENS.SONAR.mintAddress,
    marketInfoAccount: new PublicKey2(
      "4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"
    ),
    marketOwner: new PublicKey2("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
    lendingMarketAuthority: new PublicKey2(
      "sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"
    ),
    lendingPoolInfoAccount: new PublicKey2(
      "3USSy3sAZSCnUiCZJaTFt5hCHWSw87kg8tNPaEZh1p9D"
    ),
    lendingPoolTknAccount: new PublicKey2(
      "2F2yrLZQa3av7PxW7vdqiYBV4AwTbkCdswgN2BYGgtvs"
    ),
    lendingPoolFeeAccount: new PublicKey2(
      "2F2yrLZQa3av7PxW7vdqiYBV4AwTbkCdswgN2BYGgtvs"
    ),
    lendingPoolShareMint: new PublicKey2(
      "FvHrhBTTgo7q8uDX7gN5f7YzSXMrf3mwYtkay5NvkJzg"
    ),
    lendingPoolShareAccount: new PublicKey2(
      "DPMixAHcMsryStgytYBhvvtR2WbvyRxkMwG147NmXEPS"
    ),
    lendingPoolCreditMint: new PublicKey2(
      "HsKN1EC5zcArg1yoq1tMg7JKfxPW9m644mqCphrbLMC"
    ),
    lendingPoolCreditAccount: new PublicKey2(
      "7zqtZ8LhBYUUXoDCwDFcdLgNkYo8tZZqTdoFJqAL5ie8"
    )
  },
  DFL: {
    lookupTableAddress: new PublicKey2(
      "2Q144LX9XXacpEwNMFMESnuZhCNpPrABYun3AZJxnYHB"
    ),
    programId: lendProgramId,
    tokenMint: TOKENS.DFL.mintAddress,
    marketInfoAccount: new PublicKey2(
      "4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"
    ),
    marketOwner: new PublicKey2("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
    lendingMarketAuthority: new PublicKey2(
      "sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"
    ),
    lendingPoolInfoAccount: new PublicKey2(
      "74fWq2xnHBP1W6QhLDazCZAuZvp9XY3ck1PGNJg9hfhi"
    ),
    lendingPoolTknAccount: new PublicKey2(
      "5QNHLaujxMgFKYaJM55uYb3jyMGgvspizg7S5eVAD3d8"
    ),
    lendingPoolFeeAccount: new PublicKey2(
      "5QNHLaujxMgFKYaJM55uYb3jyMGgvspizg7S5eVAD3d8"
    ),
    lendingPoolShareMint: new PublicKey2(
      "CGhMaGeVLxxQGJh6Y2bEYfLazumuFyEXpxF1UgrcMRJi"
    ),
    lendingPoolShareAccount: new PublicKey2(
      "EHUF88xvJZyXzosAQ7Jzx5ozdvxCfZuwppSEo5b9pe2S"
    ),
    lendingPoolCreditMint: new PublicKey2(
      "26Eo8VcgDuNZSn8x5infdYxDKs7mYZj4JKxLUCkKP3DV"
    ),
    lendingPoolCreditAccount: new PublicKey2(
      "GYZgKQmYZBosdS3PsSRv83CdPFAtTRXRZaqdFVUnumiw"
    )
  },
  wbWBNB: {
    lookupTableAddress: new PublicKey2(
      "AJMcPqmPMnPonDKCUHMyTF8sVCkGaHxMzQdEPKtY3QbK"
    ),
    programId: lendProgramId,
    tokenMint: TOKENS.wbWBNB.mintAddress,
    marketInfoAccount: new PublicKey2(
      "4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"
    ),
    marketOwner: new PublicKey2("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
    lendingMarketAuthority: new PublicKey2(
      "sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"
    ),
    lendingPoolInfoAccount: new PublicKey2(
      "6RnKftFywvw8sceeWUM623BYyMeW216276G4atZ23CrF"
    ),
    lendingPoolTknAccount: new PublicKey2(
      "F7SQgbgunwx7PKc9i5zXJfSNsjU9xM5P8Sb18FCUBKyf"
    ),
    lendingPoolFeeAccount: new PublicKey2(
      "F7SQgbgunwx7PKc9i5zXJfSNsjU9xM5P8Sb18FCUBKyf"
    ),
    lendingPoolShareMint: new PublicKey2(
      "8EzbLKBQ8Rest1SWUtZR6zFvYTo4cCaQ73ni8sF9vtjA"
    ),
    lendingPoolShareAccount: new PublicKey2(
      "DgBnJvW9iQT5sP5YyVHeNbRjG8Xd2d53f19a2451eXkV"
    ),
    lendingPoolCreditMint: new PublicKey2(
      "6bMAnnzp3pim8wd8XypLGuqVFQ3LMuY3vkRRse1DkZ7o"
    ),
    lendingPoolCreditAccount: new PublicKey2(
      "CFsjid6WFmsyZozJkj43ssA5rRmYa97mzvfs8eRYqBeZ"
    )
  },
  stSOL: {
    lookupTableAddress: new PublicKey2(
      "kTJ8hWwRV6vmj7feyD1ppNk7rxiuPMmhrjoSmHJ75M9"
    ),
    programId: lendProgramId,
    tokenMint: TOKENS.stSOL.mintAddress,
    marketInfoAccount: new PublicKey2(
      "4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"
    ),
    marketOwner: new PublicKey2("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
    lendingMarketAuthority: new PublicKey2(
      "sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"
    ),
    lendingPoolInfoAccount: new PublicKey2(
      "9ogtxmnUF94KaPn3dB3unT5yyu7rpzSwc5igysgqFxrU"
    ),
    lendingPoolTknAccount: new PublicKey2(
      "85pJTrAVdjHNvgCcUtefwkSe9RDKnHueyvs2uTocWmWs"
    ),
    lendingPoolFeeAccount: new PublicKey2(
      "85pJTrAVdjHNvgCcUtefwkSe9RDKnHueyvs2uTocWmWs"
    ),
    lendingPoolShareMint: new PublicKey2(
      "HyxfDg47HKS1rQUXWvJH2XQhwEJM3AqrHU7sK9bSNt5h"
    ),
    lendingPoolShareAccount: new PublicKey2(
      "3H64RCjRcHEJAJK9mZ3q1mBZJ2hWuv3tUvXVY7XgFL5L"
    ),
    lendingPoolCreditMint: new PublicKey2(
      "Hhh5mQQMuWjmyUNyYFDtuXKFp9wptPuLYvH9x98FSDq8"
    ),
    lendingPoolCreditAccount: new PublicKey2(
      "CRorEh6T7aXamCsjggckCMgF1ewVavo6CfXTPNkipson"
    )
  },
  SHDW: {
    lookupTableAddress: new PublicKey2(
      "37Fuw9zCDDaohHxY12cw6viNJ4J4qZysRYQynQ75Fo9F"
    ),
    programId: lendProgramId,
    tokenMint: TOKENS.SHDW.mintAddress,
    marketInfoAccount: new PublicKey2(
      "4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"
    ),
    marketOwner: new PublicKey2("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
    lendingPoolInfoAccount: new PublicKey2(
      "8NvPwUxtBohPg45bAmzzB45qLajpfptMVgqgu5vmCNgi"
    ),
    lendingMarketAuthority: new PublicKey2(
      "sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"
    ),
    lendingPoolTknAccount: new PublicKey2(
      "EoVqoTtczavGmATCagSQU11c74MkL8enp96fGkQz6uqK"
    ),
    lendingPoolShareMint: new PublicKey2(
      "AJu1s738dGsZ8mV2XKTqjuMiAiqNGaRsixTR4Czx4mJ2"
    ),
    lendingPoolShareAccount: new PublicKey2(
      "48YJFjo3zLGSkUDXpa8mte3EMFNSPvy6pE9wUhoPSLnH"
    ),
    lendingPoolCreditMint: new PublicKey2(
      "3fu3y5yGbBPqDpwKjN1PsVurm4af6uQnpkkuSo5SxZQa"
    ),
    lendingPoolCreditAccount: new PublicKey2(
      "2zxQFrSGMSWbHMzMjxDN8ceqgbJVUyyEEwR75R5zzh4M"
    ),
    lendingPoolFeeAccount: new PublicKey2(
      "3H3QXT9oEG5DaoWfQ727jgJdhVewgrTNc8jsGgBsXvQE"
    )
  },
  BASIS: {
    lookupTableAddress: new PublicKey2(
      "CweY4HW2y4mQQmduqQNBPRHLNcsLmNV3aUNSE4EHumhe"
    ),
    programId: lendProgramId,
    tokenMint: TOKENS.BASIS.mintAddress,
    marketInfoAccount: new PublicKey2(
      "4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"
    ),
    marketOwner: new PublicKey2("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
    lendingPoolInfoAccount: new PublicKey2(
      "499SnZR7dFzLU6BF9v9obfSCsmgui3FBtYtFDakD89zQ"
    ),
    lendingMarketAuthority: new PublicKey2(
      "sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"
    ),
    lendingPoolTknAccount: new PublicKey2(
      "E85Eyi4NmBp2Der1ntoNNx3AH5FHmjp4gGoUPW7qYVJ8"
    ),
    lendingPoolShareMint: new PublicKey2(
      "9s2iRZBzSNCNTUtuzSRHcv3q4Q9SaFFhSFZVnFXGtHUp"
    ),
    lendingPoolShareAccount: new PublicKey2(
      "5phQAZCwdfbYqj1tsv4xYWREpLZuxhqnMaZ126RAfVNb"
    ),
    lendingPoolCreditMint: new PublicKey2(
      "AVRoCxDDfx525L1XJJ5JbKrPPbuHs1JL6Z17vWS3HpKc"
    ),
    lendingPoolCreditAccount: new PublicKey2(
      "FeMzBdYfGJtqeRAadiC4fStY13cjxe6GutTyFCFsP1vL"
    ),
    lendingPoolFeeAccount: new PublicKey2(
      "CtdMN3hrU9yq2GJpojbU1oQZbMXJyuVZYMfNkAPbgaZN"
    )
  },
  wUST: {
    lookupTableAddress: new PublicKey2(
      "HbSUXs2y1nMm1rpWfxeD25ZsfXEk19PuvWTZhdzvwUvo"
    ),
    programId: lendProgramId,
    tokenMint: TOKENS.wUST.mintAddress,
    marketInfoAccount: new PublicKey2(
      "4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"
    ),
    marketOwner: new PublicKey2("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
    lendingPoolInfoAccount: new PublicKey2(
      "G1PYcFc56DQHDjSyt8zRKcnxbBrKJq4Vehurzrt2KUzs"
    ),
    lendingMarketAuthority: new PublicKey2(
      "sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"
    ),
    lendingPoolTknAccount: new PublicKey2(
      "DKYpvhtPyU9yWkGZx5VM4fA1isLwpcmQ8ttqhPJgMMrJ"
    ),
    lendingPoolShareMint: new PublicKey2(
      "FPnFwW1ASFLBReVz1EPWVGkbrsnrekVkEcQnrBRkvDXS"
    ),
    lendingPoolShareAccount: new PublicKey2(
      "MeLTPseZyHm2aWSk9Gda9d6xWM8DS5QySFKBS3Vyzsv"
    ),
    lendingPoolCreditMint: new PublicKey2(
      "AbPJaMzRetUebzRV3mtoXtyPDVwEBEHif3EibVKhfcTs"
    ),
    lendingPoolCreditAccount: new PublicKey2(
      "mYi2FARcy26GxkDuTneTybM75izMAR7swioxUMvyPRa"
    ),
    lendingPoolFeeAccount: new PublicKey2(
      "5wTjKzJyEJHbjw4fsZM51fKBDwiq5JSA3bSzzRE2PXob"
    )
  },
  GST: {
    lookupTableAddress: new PublicKey2(
      "GBAgJPNhEmt3JwubXVgd66zoJZP8XBGD7FwsYzo8WRkh"
    ),
    programId: lendProgramId,
    tokenMint: TOKENS.GST.mintAddress,
    marketInfoAccount: new PublicKey2(
      "4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"
    ),
    marketOwner: new PublicKey2("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
    lendingMarketAuthority: new PublicKey2(
      "sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"
    ),
    lendingPoolInfoAccount: new PublicKey2(
      "93zxnt1gPepvUjWYPP32uANnhAK6NNphGD6A2yQFhk4s"
    ),
    lendingPoolTknAccount: new PublicKey2(
      "4Y7VyaAWEDaQSkeYXbXRf1d82ZPViVRZ7Dt4fjHKgotL"
    ),
    lendingPoolShareMint: new PublicKey2(
      "4GDuSZwHUQvL6zga2UNdjgXSqXjcWaUYc4S36N9q4567"
    ),
    lendingPoolShareAccount: new PublicKey2(
      "GjmnPLwAAA5aZCfN5sCwcbartonu8iyVMPqSnfkyvtC9"
    ),
    lendingPoolCreditMint: new PublicKey2(
      "8Nq8YV7BYoe3dqdTX996TRDB2iYryAHbzZTXntt2A3Uo"
    ),
    lendingPoolCreditAccount: new PublicKey2(
      "8dzJgZxp7rBEbKhvFYXTMXKLmkqscn9UXwzcGA14fEgN"
    ),
    lendingPoolFeeAccount: new PublicKey2(
      "ERyCAJh59hn8187Kn6wiKpeXdnGw4hptrbn55vh9YJQ1"
    )
  },
  GMT: {
    lookupTableAddress: new PublicKey2(
      "3KzviRz2snQevukX9FWt5n72XQFX48czkWJPsAiGoJZz"
    ),
    programId: lendProgramId,
    tokenMint: TOKENS.GMT.mintAddress,
    marketInfoAccount: new PublicKey2(
      "4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"
    ),
    lendingMarketAuthority: new PublicKey2(
      "sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"
    ),
    marketOwner: new PublicKey2("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
    lendingPoolInfoAccount: new PublicKey2(
      "9b5QWW2SqjmhRxtYTHXj6LFkmUkahrLAMQer3rEurd1k"
    ),
    lendingPoolTknAccount: new PublicKey2(
      "7ZosDG7JWwXFnzhQ2aQ1ghCSQ9mrQzvH3ZhduhRPHgJB"
    ),
    lendingPoolShareMint: new PublicKey2(
      "9R6toP3xTFBWoqwBY8bfjUj6Tyd5hyR192jY4NeMmkNg"
    ),
    lendingPoolShareAccount: new PublicKey2(
      "BPPtMXgGkpo5HuDbiFyQWiA5r4BfxRWfwUA8XoKPixb7"
    ),
    lendingPoolCreditMint: new PublicKey2(
      "B1yWMWp7nvSV2upJxFUxpUkDiZAVD2MabnRgaBtR14Dm"
    ),
    lendingPoolCreditAccount: new PublicKey2(
      "6NvFHm9p6oSK8DJ9jurTMkC8KBo2hCk8UQm19Nub7R6X"
    ),
    lendingPoolFeeAccount: new PublicKey2(
      "ejLn4fFkVS29aYksC87TQYwcu8wy1tvc2AySpfHGabQ"
    )
  },
  ZBC: {
    lookupTableAddress: new PublicKey2(
      "FNVsWfdmEcAjdxUZtRC9165RUt8zQh3k3mANCQL4sRAu"
    ),
    programId: lendProgramId,
    tokenMint: TOKENS.ZBC.mintAddress,
    marketInfoAccount: new PublicKey2(
      "4XNif294wbrxj6tJ8K5Rg7SuaEACnu9s2L27i28MQB6E"
    ),
    lendingMarketAuthority: new PublicKey2(
      "sCDiYj7X7JmXg5fVq2nqED2q1Wqjo7PnqMgH3casMem"
    ),
    marketOwner: new PublicKey2("7MBLg6oV5phip11YBbJPuq7u38kdzSi9PM3BifKSpLaR"),
    lendingPoolInfoAccount: new PublicKey2(
      "6uPerEUyomCC5qn4z9xF3fbVa7SCytcKoCVannPDez5R"
    ),
    lendingPoolTknAccount: new PublicKey2(
      "7quTFi1N9EZfz7jUx7ui9wx9rk1rb3F1Enjbh8Me2tSs"
    ),
    lendingPoolShareMint: new PublicKey2(
      "HUry6RSbDasMfU7LXZ5QR42YajkbMuLBQkyV742xgFKx"
    ),
    lendingPoolShareAccount: new PublicKey2(
      "BHuVdzrDTssFdWTqt15fsdJzJuHESQrMhSnc3jsJ8X2Y"
    ),
    lendingPoolCreditMint: new PublicKey2(
      "ADWAgFjannL985LDKQsQ9B833BFDMj4q8qQappe2F3X9"
    ),
    lendingPoolCreditAccount: new PublicKey2(
      "5xqkszmojpTFy3wGQTXFB7XoLyWXxw7ymmHN9iNYVdUK"
    ),
    lendingPoolFeeAccount: new PublicKey2(
      "qt8KMibr5UcB1WZGVdgyFNGvmzEtp3faUqfcKDytLzq"
    )
  }
};

// src/index.ts
var LENDING_CONFIG = lendingPools;
export {
  LENDING_CONFIG
};
