// @dradex/idl/dist/cjs/layout/queue.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventQueue = exports.EVENT_QUEUE_LAYOUT = void 0;
const buffer_layout_1 = require("@solana/buffer-layout");
const base_1 = require("./base");
const EVENT_ITEM_LAYOUT = (0, buffer_layout_1.struct)([
    (0, buffer_layout_1.u8)("flag"),
    (0, buffer_layout_1.u8)("fee_tier"),
    (0, buffer_layout_1.seq)((0, buffer_layout_1.u8)(), 6),
    (0, base_1.u128)("key"),
    (0, base_1.publicKeyLayout)("owner"),
    (0, base_1.u64)("quantity"),
    (0, base_1.u64)("total"),
]);
const QUEUE_LEN = (0, buffer_layout_1.u32)("len");
exports.EVENT_QUEUE_LAYOUT = (0, buffer_layout_1.struct)([
    (0, base_1.u64)("flag"),
    QUEUE_LEN,
    (0, buffer_layout_1.seq)((0, buffer_layout_1.u8)(), 4),
    (0, buffer_layout_1.seq)(EVENT_ITEM_LAYOUT, (0, buffer_layout_1.offset)(QUEUE_LEN, -8), "items"),
]);
class EventQueue {
    constructor(len, items) {
        this.len = len;
        this.items = items;
    }
    allItems() {
        return this.items;
    }
    get length() {
        return this.len;
    }
    static decode(buffer) {
        const { len, items } = exports.EVENT_QUEUE_LAYOUT.decode(buffer);
        return new EventQueue(len, items);
    }
}
exports.EventQueue = EventQueue;
(0, base_1.setLayoutDecoder)(exports.EVENT_QUEUE_LAYOUT, ({ len, items }) => new EventQueue(len, items));
