// @quarryprotocol/quarry-sdk/dist/esm/wrappers/redeemer/index.js
import { __awaiter } from "tslib";
import { TransactionEnvelope } from "@saberhq/solana-contrib";
import { getATAAddress, getOrCreateATA, TOKEN_PROGRAM_ID, } from "@saberhq/token-utils";
import { SystemProgram } from "@solana/web3.js";
import { findRedeemerKey } from "./pda";
export * from "./pda";
export class RedeemerWrapper {
    constructor(sdk, iouMint, redemptionMint, key, data) {
        this.sdk = sdk;
        this.iouMint = iouMint;
        this.redemptionMint = redemptionMint;
        this.key = key;
        this.data = data;
    }
    get program() {
        return this.sdk.programs.Redeemer;
    }
    static load({ sdk, iouMint, redemptionMint, }) {
        return __awaiter(this, void 0, void 0, function* () {
            const [redeemer] = yield findRedeemerKey({ iouMint, redemptionMint });
            const program = sdk.programs.Redeemer;
            const data = yield program.account.redeemer.fetch(redeemer);
            return new RedeemerWrapper(sdk, iouMint, redemptionMint, redeemer, data);
        });
    }
    static createRedeemer({ sdk, iouMint, redemptionMint, }) {
        return __awaiter(this, void 0, void 0, function* () {
            const { provider } = sdk;
            const [redeemer, bump] = yield findRedeemerKey({ iouMint, redemptionMint });
            const ata = yield getOrCreateATA({
                provider,
                mint: redemptionMint,
                owner: redeemer,
            });
            return {
                bump,
                vaultTokenAccount: ata.address,
                tx: new TransactionEnvelope(sdk.provider, [
                    ...(ata.instruction ? [ata.instruction] : []),
                    sdk.programs.Redeemer.instruction.createRedeemer(bump, {
                        accounts: {
                            redeemer,
                            iouMint,
                            redemptionMint,
                            payer: provider.wallet.publicKey,
                            systemProgram: SystemProgram.programId,
                        },
                    }),
                ]),
            };
        });
    }
    /**
     * redeemTokensIx
     */
    redeemTokensIx(args) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.program.instruction.redeemTokens(args.tokenAmount, {
                accounts: yield this.getRedeemTokenAccounts(args),
            });
        });
    }
    redeemTokens(args) {
        return __awaiter(this, void 0, void 0, function* () {
            return new TransactionEnvelope(this.sdk.provider, [
                yield this.redeemTokensIx(args),
            ]);
        });
    }
    getVaultAddress() {
        return __awaiter(this, void 0, void 0, function* () {
            return yield getATAAddress({
                mint: this.redemptionMint,
                owner: this.key,
            });
        });
    }
    getRedeemTokenAccounts(args) {
        return __awaiter(this, void 0, void 0, function* () {
            const { iouSource, redemptionDestination, sourceAuthority } = args;
            return {
                redeemer: this.key,
                iouMint: this.data.iouMint,
                redemptionMint: this.data.redemptionMint,
                redemptionVault: yield getATAAddress({
                    mint: this.data.redemptionMint,
                    owner: this.key,
                }),
                tokenProgram: TOKEN_PROGRAM_ID,
                sourceAuthority,
                iouSource,
                redemptionDestination,
            };
        });
    }
}
