// @quarryprotocol/quarry-sdk/dist/cjs/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.QUARRY_FEE_SETTER = exports.QUARRY_FEE_TO = exports.QUARRY_CODERS = exports.QUARRY_IDLS = exports.QUARRY_ADDRESSES = void 0;
const anchor_contrib_1 = require("@saberhq/anchor-contrib");
const web3_js_1 = require("@solana/web3.js");
const quarry_mine_1 = require("./idls/quarry_mine");
const quarry_mint_wrapper_1 = require("./idls/quarry_mint_wrapper");
const quarry_redeemer_1 = require("./idls/quarry_redeemer");
const programs_1 = require("./programs");
const registry_1 = require("./programs/registry");
/**
 * Quarry program addresses.
 */
exports.QUARRY_ADDRESSES = {
    MergeMine: new web3_js_1.PublicKey("QMMD16kjauP5knBwxNUJRZ1Z5o3deBuFrqVjBVmmqto"),
    Mine: new web3_js_1.PublicKey("QMNeHCGYnLVDn1icRAfQZpjPLBNkfGbSKRB83G5d8KB"),
    MintWrapper: new web3_js_1.PublicKey("QMWoBmAyJLAsA1Lh9ugMTw2gciTihncciphzdNzdZYV"),
    Operator: new web3_js_1.PublicKey("QoP6NfrQbaGnccXQrMLUkog2tQZ4C1RFgJcwDnT8Kmz"),
    Redeemer: new web3_js_1.PublicKey("QRDxhMw1P2NEfiw5mYXG79bwfgHTdasY2xNP76XSea9"),
    Registry: new web3_js_1.PublicKey("QREGBnEj9Sa5uR91AV8u3FxThgP5ZCvdZUW2bHAkfNc"),
};
/**
 * Quarry program IDLs.
 */
exports.QUARRY_IDLS = {
    MergeMine: programs_1.QuarryMergeMineJSON,
    Mine: quarry_mine_1.QuarryMineJSON,
    MintWrapper: quarry_mint_wrapper_1.QuarryMintWrapperJSON,
    Operator: programs_1.QuarryOperatorJSON,
    Redeemer: quarry_redeemer_1.QuarryRedeemerJSON,
    Registry: registry_1.QuarryRegistryJSON,
};
/**
 * Quarry program IDLs.
 */
exports.QUARRY_CODERS = (0, anchor_contrib_1.buildCoderMap)(exports.QUARRY_IDLS, exports.QUARRY_ADDRESSES);
/**
 * Recipient of protocol fees.
 */
exports.QUARRY_FEE_TO = new web3_js_1.PublicKey("4MMZH3ih1aSty2nx4MC3kSR94Zb55XsXnqb5jfEcyHWQ");
/**
 * Sets the protocol fees.
 */
exports.QUARRY_FEE_SETTER = new web3_js_1.PublicKey("4MMZH3ih1aSty2nx4MC3kSR94Zb55XsXnqb5jfEcyHWQ");
