// @mercurial-finance/mango-v4/dist/chunk-S6N5Y2X2.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true});// ts/client/src/constants/index.ts
var _anchor = require('@project-serum/anchor');
var _web3js = require('@solana/web3.js');
var RUST_U64_MAX = () => {
  return new (0, _anchor.BN)("18446744073709551615");
};
var RUST_I64_MAX = () => {
  return new (0, _anchor.BN)("9223372036854775807");
};
var RUST_I64_MIN = () => {
  return new (0, _anchor.BN)("-9223372036854775807");
};
var OPENBOOK_PROGRAM_ID = {
  devnet: new (0, _web3js.PublicKey)("EoTcMgcDRTJVZDMZWBoU6rhYHZfkNTVEAfz3uUJRcYGj"),
  "mainnet-beta": new (0, _web3js.PublicKey)("srmqPvymJeFKQ4zGQed1GFppgkRHL9kaELCbyksJtPX")
};
var MANGO_V4_ID = {
  testnet: new (0, _web3js.PublicKey)("4MangoMjqJ2firMokCjjGgoK8d4MXcrgL7XJaL3w6fVg"),
  devnet: new (0, _web3js.PublicKey)("4MangoMjqJ2firMokCjjGgoK8d4MXcrgL7XJaL3w6fVg"),
  "mainnet-beta": new (0, _web3js.PublicKey)("4MangoMjqJ2firMokCjjGgoK8d4MXcrgL7XJaL3w6fVg")
};







exports.RUST_U64_MAX = RUST_U64_MAX; exports.RUST_I64_MAX = RUST_I64_MAX; exports.RUST_I64_MIN = RUST_I64_MIN; exports.OPENBOOK_PROGRAM_ID = OPENBOOK_PROGRAM_ID; exports.MANGO_V4_ID = MANGO_V4_ID;
