// @mercurial-finance/mango-v4/dist/chunk-UGHO4YHY.js
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _optionalChain(ops) { let lastAccessLHS = undefined; let value = ops[0]; let i = 1; while (i < ops.length) { const op = ops[i]; const fn = ops[i + 1]; i += 2; if ((op === 'optionalAccess' || op === 'optionalCall') && value == null) { return undefined; } if (op === 'access' || op === 'optionalAccess') { lastAccessLHS = value; value = fn(value); } else if (op === 'call' || op === 'optionalCall') { value = fn((...args) => value.call(lastAccessLHS, ...args)); lastAccessLHS = undefined; } } return value; }

var _chunkS6N5Y2X2js = require('./chunk-S6N5Y2X2.js');




var _chunkS3PQ6OXSjs = require('./chunk-S3PQ6OXS.js');

// ts/client/src/accounts/serum3.ts
var _bytes = require('@project-serum/anchor/dist/cjs/utils/bytes');
var _web3js = require('@solana/web3.js');
var Serum3Market = class {
  constructor(publicKey, group, baseTokenIndex, quoteTokenIndex, name, serumProgram, serumMarketExternal, marketIndex, registrationTime, reduceOnly) {
    this.publicKey = publicKey;
    this.group = group;
    this.baseTokenIndex = baseTokenIndex;
    this.quoteTokenIndex = quoteTokenIndex;
    this.serumProgram = serumProgram;
    this.serumMarketExternal = serumMarketExternal;
    this.marketIndex = marketIndex;
    this.registrationTime = registrationTime;
    this.reduceOnly = reduceOnly;
    this.name = _bytes.utf8.decode(new Uint8Array(name)).split("\0")[0];
  }
  static from(publicKey, obj) {
    return new Serum3Market(
      publicKey,
      obj.group,
      obj.baseTokenIndex,
      obj.quoteTokenIndex,
      obj.name,
      obj.serumProgram,
      obj.serumMarketExternal,
      obj.marketIndex,
      obj.registrationTime,
      obj.reduceOnly == 1
    );
  }
  async findOoPda(programId, mangoAccount) {
    const [openOrderPublicKey] = await _web3js.PublicKey.findProgramAddress(
      [
        Buffer.from("Serum3OO"),
        mangoAccount.toBuffer(),
        this.publicKey.toBuffer()
      ],
      programId
    );
    return openOrderPublicKey;
  }
  getFeeRates(taker = true) {
    const ratesBps = this.name === "USDT/USDC" ? { maker: -0.5, taker: 1 } : { maker: -2, taker: 4 };
    return taker ? ratesBps.maker * 1e-4 : ratesBps.taker * 1e-4;
  }
  /**
   *
   * @param group
   * @returns maximum leverage one can bid on this market, this is only for display purposes,
   *  also see getMaxQuoteForSerum3BidUi and getMaxBaseForSerum3AskUi
   */
  maxBidLeverage(group) {
    const baseBank = group.getFirstBankByTokenIndex(this.baseTokenIndex);
    const quoteBank = group.getFirstBankByTokenIndex(this.quoteTokenIndex);
    if (quoteBank.initLiabWeight.sub(baseBank.initAssetWeight).lte(_chunkS3PQ6OXSjs.ZERO_I80F48.call(void 0, ))) {
      return _chunkS3PQ6OXSjs.MAX_I80F48.call(void 0, ).toNumber();
    }
    return _chunkS3PQ6OXSjs.ONE_I80F48.call(void 0, ).div(quoteBank.initLiabWeight.sub(baseBank.initAssetWeight)).toNumber();
  }
  /**
   *
   * @param group
   * @returns maximum leverage one can ask on this market, this is only for display purposes,
   *  also see getMaxQuoteForSerum3BidUi and getMaxBaseForSerum3AskUi
   */
  maxAskLeverage(group) {
    const baseBank = group.getFirstBankByTokenIndex(this.baseTokenIndex);
    const quoteBank = group.getFirstBankByTokenIndex(this.quoteTokenIndex);
    if (baseBank.initLiabWeight.sub(quoteBank.initAssetWeight).lte(_chunkS3PQ6OXSjs.ZERO_I80F48.call(void 0, ))) {
      return _chunkS3PQ6OXSjs.MAX_I80F48.call(void 0, ).toNumber();
    }
    return _chunkS3PQ6OXSjs.ONE_I80F48.call(void 0, ).div(baseBank.initLiabWeight.sub(quoteBank.initAssetWeight)).toNumber();
  }
  async loadBids(client, group) {
    const serum3MarketExternal = group.getSerum3ExternalMarket(
      this.serumMarketExternal
    );
    return await serum3MarketExternal.loadBids(
      client.program.provider.connection
    );
  }
  async loadAsks(client, group) {
    const serum3MarketExternal = group.getSerum3ExternalMarket(
      this.serumMarketExternal
    );
    return await serum3MarketExternal.loadAsks(
      client.program.provider.connection
    );
  }
  async logOb(client, group) {
    let res = ``;
    res += `  ${this.name} OrderBook`;
    let orders = await _optionalChain([this, 'optionalAccess', _ => _.loadAsks, 'call', _2 => _2(client, group)]);
    for (const order of orders.items(true)) {
      res += `
  ${order.price.toString().padStart(10)}, ${order.size.toString().padStart(10)}`;
    }
    res += `
  --------------------------`;
    orders = await _optionalChain([this, 'optionalAccess', _3 => _3.loadBids, 'call', _4 => _4(client, group)]);
    for (const order of orders.items(true)) {
      res += `
  ${order.price.toString().padStart(10)}, ${order.size.toString().padStart(10)}`;
    }
    return res;
  }
};
var Serum3SelfTradeBehavior = class {
};
Serum3SelfTradeBehavior.decrementTake = { decrementTake: {} };
Serum3SelfTradeBehavior.cancelProvide = { cancelProvide: {} };
Serum3SelfTradeBehavior.abortTransaction = { abortTransaction: {} };
var Serum3OrderType = class {
};
Serum3OrderType.limit = { limit: {} };
Serum3OrderType.immediateOrCancel = { immediateOrCancel: {} };
Serum3OrderType.postOnly = { postOnly: {} };
var Serum3Side = class {
};
Serum3Side.bid = { bid: {} };
Serum3Side.ask = { ask: {} };
async function generateSerum3MarketExternalVaultSignerAddress(cluster, serum3Market, serum3MarketExternal) {
  return await _web3js.PublicKey.createProgramAddress(
    [
      serum3Market.serumMarketExternal.toBuffer(),
      serum3MarketExternal.decoded.vaultSignerNonce.toArrayLike(
        Buffer,
        "le",
        8
      )
    ],
    _chunkS6N5Y2X2js.OPENBOOK_PROGRAM_ID[cluster]
  );
}







exports.Serum3Market = Serum3Market; exports.Serum3SelfTradeBehavior = Serum3SelfTradeBehavior; exports.Serum3OrderType = Serum3OrderType; exports.Serum3Side = Serum3Side; exports.generateSerum3MarketExternalVaultSignerAddress = generateSerum3MarketExternalVaultSignerAddress;
