// @mercurial-finance/vault-sdk/dist/cjs/src/vault/strategy/apricotWithoutLM.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const web3_js_1 = require("@solana/web3.js");
const spl_token_1 = require("@solana/spl-token");
const apricot = __importStar(require("@mercurial-finance/apricot-sdk"));
const anchor = __importStar(require("@project-serum/anchor"));
const constants_1 = require("../constants");
class ApricotWithoutLMHandler {
    constructor() {
        this.address = new apricot.Addresses(apricot.PUBLIC_CONFIG);
    }
    withdraw(walletPubKey, program, strategy, vault, tokenVault, vaultState, userToken, userLp, amount, preInstructions, postInstructions, opt) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!walletPubKey)
                throw new Error('No user wallet public key');
            const [[collateralVault], [userInfoSignerPda]] = yield Promise.all([constants_1.SEEDS.COLLATERAL_VAULT_PREFIX, constants_1.SEEDS.APRICOT_USER_INFO_SIGNER_PREFIX].map((seed) => __awaiter(this, void 0, void 0, function* () {
                return yield web3_js_1.PublicKey.findProgramAddress([Buffer.from(seed), new web3_js_1.PublicKey(strategy.pubkey).toBuffer()], program.programId);
            })));
            const [basePda] = yield this.address.getBasePda();
            const [userInfo, assetPoolSpl, poolSummaries, priceSummaries, userPageStats] = yield Promise.all([
                this.address.getUserInfoKey(userInfoSignerPda),
                this.address.getAssetPoolSplKey(basePda, vaultState.tokenMint.toBase58()),
                this.address.getPoolSummariesKey(),
                this.address.getPriceSummariesKey(basePda),
                this.address.getUserPagesStatsKey(),
            ]);
            const accounts = [
                { pubkey: userInfo },
                { pubkey: assetPoolSpl },
                { pubkey: poolSummaries },
                { pubkey: priceSummaries },
                { pubkey: userInfoSignerPda },
                { pubkey: basePda },
                { pubkey: userPageStats },
            ];
            const remainingAccounts = accounts.map((account) => ({
                pubkey: account.pubkey,
                isWritable: true,
                isSigner: false,
            }));
            const txAccounts = {
                vault,
                strategy: strategy.pubkey,
                reserve: new web3_js_1.PublicKey(strategy.state.reserve),
                strategyProgram: this.address.getProgramKey(),
                collateralVault,
                tokenVault,
                feeVault: vaultState.feeVault,
                userToken,
                userLp,
                tokenProgram: spl_token_1.TOKEN_PROGRAM_ID,
            };
            if (opt === null || opt === void 0 ? void 0 : opt.affiliate) {
                const tx = yield opt.affiliate.affiliateProgram.methods
                    .withdrawDirectlyFromStrategy(new anchor.BN(amount), new anchor.BN(0))
                    .accounts(Object.assign(Object.assign({}, txAccounts), { partner: opt.affiliate.partner, user: opt.affiliate.user, vaultProgram: program.programId, vaultLpMint: vaultState.lpMint, owner: walletPubKey }))
                    .remainingAccounts(remainingAccounts)
                    .preInstructions(preInstructions.concat(
                // refreshUser Instruction
                new web3_js_1.TransactionInstruction({
                    programId: this.address.getProgramKey(),
                    keys: [
                        { pubkey: userInfoSignerPda, isSigner: false, isWritable: false },
                        { pubkey: userInfo, isSigner: false, isWritable: true },
                        { pubkey: poolSummaries, isSigner: false, isWritable: false },
                    ],
                    data: Buffer.from([apricot.CMD_REFRESH_USER]),
                })))
                    .postInstructions(postInstructions)
                    .transaction();
                return tx;
            }
            const tx = yield program.methods
                .withdrawDirectlyFromStrategy(new anchor.BN(amount), new anchor.BN(0))
                .accounts(Object.assign(Object.assign({}, txAccounts), { lpMint: vaultState.lpMint, user: walletPubKey }))
                .remainingAccounts(remainingAccounts)
                .preInstructions(preInstructions.concat(
            // refreshUser Instruction
            new web3_js_1.TransactionInstruction({
                programId: this.address.getProgramKey(),
                keys: [
                    { pubkey: userInfoSignerPda, isSigner: false, isWritable: false },
                    { pubkey: userInfo, isSigner: false, isWritable: true },
                    { pubkey: poolSummaries, isSigner: false, isWritable: false },
                ],
                data: Buffer.from([apricot.CMD_REFRESH_USER]),
            })))
                .postInstructions(postInstructions)
                .transaction();
            return tx;
        });
    }
}
exports.default = ApricotWithoutLMHandler;
