// @mercurial-finance/vault-sdk/dist/cjs/src/vault/strategy/cypher.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const web3_js_1 = require("@solana/web3.js");
const cypher_client_1 = require("@mercurial-finance/cypher-client");
const spl_token_1 = require("@solana/spl-token");
const bn_js_1 = __importDefault(require("bn.js"));
const constants_1 = require("../constants");
const utils_1 = require("../utils");
class CypherHandler {
    constructor(cluster, program) {
        this.cluster = cluster;
        this.program = program;
        this.cypherClient = new cypher_client_1.CypherClient(cluster, program.provider.connection.rpcEndpoint);
    }
    withdraw(walletPubKey, program, strategy, vault, tokenVault, vaultState, userToken, userLp, amount, preInstructions, postInstructions, opt) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!walletPubKey)
                throw new Error('No user wallet public key');
            const strategyBuffer = new web3_js_1.PublicKey(strategy.pubkey).toBuffer();
            const [collateralVault] = yield web3_js_1.PublicKey.findProgramAddress([Buffer.from(constants_1.SEEDS.COLLATERAL_VAULT_PREFIX), strategyBuffer], program.programId);
            const [strategyOwnerPubkey] = yield web3_js_1.PublicKey.findProgramAddress([Buffer.from(constants_1.SEEDS.CYPHER), strategyBuffer], program.programId);
            const [strategyOwnerATA] = yield (0, utils_1.getOrCreateATAInstruction)(vaultState.tokenMint, strategyOwnerPubkey, this.program.provider.connection);
            const [clearingPubkey] = (0, cypher_client_1.derivePublicClearingAddress)(this.cypherClient.cypherPID);
            const [masterAccPubkey] = (0, cypher_client_1.deriveAccountAddress)(strategyOwnerPubkey, 0, this.cypherClient.cypherPID);
            const [subAccPubkey] = (0, cypher_client_1.deriveSubAccountAddress)(masterAccPubkey, 0, this.cypherClient.cypherPID);
            const [poolNodePubKey] = (0, cypher_client_1.derivePoolNodeAddress)(strategy.state.reserve, 0, this.cypherClient.cypherPID);
            const [poolTokenVaultPubkey] = (0, cypher_client_1.derivePoolNodeVaultAddress)(poolNodePubKey, this.cypherClient.cypherPID);
            const [vaultSigner] = (0, cypher_client_1.derivePoolNodeVaultSigner)(poolNodePubKey, this.cypherClient.cypherPID);
            const accounts = [
                { pubkey: clearingPubkey },
                { pubkey: cypher_client_1.CONFIGS[this.cluster].CACHE },
                { pubkey: masterAccPubkey, isWritable: true },
                { pubkey: subAccPubkey, isWritable: true },
                { pubkey: poolNodePubKey, isWritable: true },
                { pubkey: poolTokenVaultPubkey, isWritable: true },
                { pubkey: vaultState.tokenMint },
                { pubkey: strategyOwnerPubkey, isWritable: true },
                { pubkey: strategyOwnerATA, isWritable: true },
                { pubkey: vaultSigner, isWritable: true },
            ];
            const remainingAccounts = [];
            for (const account of accounts) {
                remainingAccounts.push({
                    pubkey: account.pubkey,
                    isWritable: !!account.isWritable,
                    isSigner: false,
                });
            }
            const txAccounts = {
                vault,
                strategy: new web3_js_1.PublicKey(strategy.pubkey),
                reserve: new web3_js_1.PublicKey(strategy.state.reserve),
                strategyProgram: this.cypherClient.cypherPID,
                collateralVault,
                feeVault: vaultState.feeVault,
                tokenVault,
                userToken,
                userLp,
                tokenProgram: spl_token_1.TOKEN_PROGRAM_ID,
            };
            if (opt === null || opt === void 0 ? void 0 : opt.affiliate) {
                const tx = yield opt.affiliate.affiliateProgram.methods
                    .withdrawDirectlyFromStrategy(new bn_js_1.default(amount), new bn_js_1.default(0))
                    .accounts(Object.assign(Object.assign({}, txAccounts), { partner: opt.affiliate.partner, user: opt.affiliate.user, vaultProgram: program.programId, vaultLpMint: vaultState.lpMint, owner: walletPubKey }))
                    .remainingAccounts(remainingAccounts)
                    .preInstructions(preInstructions)
                    .postInstructions(postInstructions)
                    .transaction();
                return tx;
            }
            const tx = yield program.methods
                .withdrawDirectlyFromStrategy(new bn_js_1.default(amount), new bn_js_1.default(0))
                .accounts(Object.assign(Object.assign({}, txAccounts), { lpMint: vaultState.lpMint, user: walletPubKey }))
                .remainingAccounts(remainingAccounts)
                .preInstructions(preInstructions)
                .postInstructions(postInstructions)
                .transaction();
            return tx;
        });
    }
}
exports.default = CypherHandler;
