// @mercurial-finance/vault-sdk/dist/cjs/src/vault/strategy/drift.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const web3_js_1 = require("@solana/web3.js");
const drift_sdk_1 = require("@mercurial-finance/drift-sdk");
const spl_token_1 = require("@solana/spl-token");
const bn_js_1 = __importDefault(require("bn.js"));
const constants_1 = require("../constants");
class DriftHandler {
    constructor(cluster, program) {
        this.sdkConfig = (0, drift_sdk_1.initialize)({ env: cluster === 'testnet' ? 'mainnet-beta' : cluster });
        this.cluster = cluster;
        this.driftClient = new drift_sdk_1.DriftClient({
            connection: program.provider.connection,
            programID: new web3_js_1.PublicKey(this.sdkConfig.DRIFT_PROGRAM_ID),
            wallet: new drift_sdk_1.Wallet(web3_js_1.Keypair.generate()),
        });
        this.programId = new web3_js_1.PublicKey(this.sdkConfig.DRIFT_PROGRAM_ID);
    }
    withdraw(walletPubKey, program, strategy, vault, tokenVault, vaultState, userToken, userLp, amount, preInstructions, postInstructions, opt) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!walletPubKey)
                throw new Error('No user wallet public key');
            yield this.driftClient.subscribe();
            const spotMarkets = this.cluster === 'devnet' ? drift_sdk_1.DevnetSpotMarkets : drift_sdk_1.MainnetSpotMarkets;
            const spotMarket = spotMarkets.find((market) => market.mint.equals(vaultState.tokenMint));
            if (!spotMarket)
                throw new Error('Spot market not found');
            yield this.driftClient.accountSubscriber.addSpotMarket(spotMarket.marketIndex);
            const strategyBuffer = new web3_js_1.PublicKey(strategy.pubkey).toBuffer();
            const [collateralVault] = yield web3_js_1.PublicKey.findProgramAddress([Buffer.from(constants_1.SEEDS.COLLATERAL_VAULT_PREFIX), strategyBuffer], program.programId);
            // subAccountId is 16 bytes, therefore we slice 2 bytes only
            const subAccountId = new bn_js_1.default(strategy.state.bumps.slice(0, 2), 'le');
            const userPubKey = (0, drift_sdk_1.getUserAccountPublicKeySync)(this.programId, vault, subAccountId.toNumber());
            const userStatsPubKey = (0, drift_sdk_1.getUserStatsAccountPublicKey)(this.programId, vault);
            const spotMarketVaultPubKey = yield (0, drift_sdk_1.getSpotMarketVaultPublicKey)(this.programId, spotMarket.marketIndex);
            const driftState = yield (0, drift_sdk_1.getDriftStateAccountPublicKey)(this.programId);
            const driftSigner = (0, drift_sdk_1.getDriftSignerPublicKey)(this.programId);
            const accounts = [
                { pubkey: userPubKey, isWritable: true },
                { pubkey: userStatsPubKey, isWritable: true },
                { pubkey: spotMarketVaultPubKey, isWritable: true },
                { pubkey: driftState },
                { pubkey: driftSigner },
            ];
            const remainingAccounts = [];
            for (const account of accounts) {
                remainingAccounts.push({
                    pubkey: account.pubkey,
                    isWritable: !!account.isWritable,
                    isSigner: false,
                });
            }
            const driftRemainingAccounts = this.driftClient.getRemainingAccounts({
                userAccounts: [],
                writableSpotMarketIndexes: [spotMarket.marketIndex],
                useMarketLastSlotCache: false,
            });
            remainingAccounts.push(...driftRemainingAccounts);
            // prevent duplicate as spot market account pubkey will be add on program side
            const remainingAccountsWithoutReserve = remainingAccounts.filter(({ pubkey }) => !pubkey.equals(strategy.state.reserve));
            const txAccounts = {
                vault,
                strategy: new web3_js_1.PublicKey(strategy.pubkey),
                reserve: new web3_js_1.PublicKey(strategy.state.reserve),
                strategyProgram: this.sdkConfig.DRIFT_PROGRAM_ID,
                collateralVault,
                feeVault: vaultState.feeVault,
                tokenVault,
                userToken,
                userLp,
                tokenProgram: spl_token_1.TOKEN_PROGRAM_ID,
            };
            if (opt === null || opt === void 0 ? void 0 : opt.affiliate) {
                const tx = yield opt.affiliate.affiliateProgram.methods
                    .withdrawDirectlyFromStrategy(new bn_js_1.default(amount), new bn_js_1.default(0))
                    .accounts(Object.assign(Object.assign({}, txAccounts), { partner: opt.affiliate.partner, user: opt.affiliate.user, vaultProgram: program.programId, vaultLpMint: vaultState.lpMint, owner: walletPubKey }))
                    .remainingAccounts(remainingAccountsWithoutReserve)
                    .preInstructions(preInstructions)
                    .postInstructions(postInstructions)
                    .transaction();
                return tx;
            }
            const tx = yield program.methods
                .withdrawDirectlyFromStrategy(new bn_js_1.default(amount), new bn_js_1.default(0))
                .accounts(Object.assign(Object.assign({}, txAccounts), { lpMint: vaultState.lpMint, user: walletPubKey }))
                .remainingAccounts(remainingAccountsWithoutReserve)
                .preInstructions(preInstructions)
                .postInstructions(postInstructions)
                .transaction();
            return tx;
        });
    }
}
exports.default = DriftHandler;
