// @mercurial-finance/vault-sdk/dist/cjs/src/vault/strategy/mango.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const mango = __importStar(require("@blockworks-foundation/mango-client"));
const anchor = __importStar(require("@project-serum/anchor"));
const spl_token_1 = require("@solana/spl-token");
const web3_js_1 = require("@solana/web3.js");
const constants_1 = require("../constants");
class MangoHandler {
    withdraw(walletPubKey, program, strategy, vault, tokenVault, vaultState, userToken, userLp, amount, preInstructions, postInstructions, opt) {
        return __awaiter(this, void 0, void 0, function* () {
            const [mangoAccountPK] = yield web3_js_1.PublicKey.findProgramAddress([MangoHandler.MangoGrouPK.toBuffer(), vault.toBuffer(), Buffer.from([0, 0, 0, 0, 0, 0, 0, 0])], MangoHandler.MangoProgramId);
            const mangoClient = new mango.MangoClient(program.provider.connection, MangoHandler.MangoProgramId);
            const rootBankPK = strategy.state.reserve;
            const mangoGroupState = yield mangoClient.getMangoGroup(MangoHandler.MangoGrouPK);
            yield mangoGroupState.loadRootBanks(program.provider.connection);
            const rootBankIdx = mangoGroupState.getRootBankIndex(new web3_js_1.PublicKey(rootBankPK));
            const rootBankState = mangoGroupState.rootBankAccounts[rootBankIdx];
            if (!rootBankState)
                throw new Error('Root bank state not found');
            const nodeBankPK = rootBankState.nodeBanks[0];
            const nodeBankState = rootBankState.nodeBankAccounts.find((t) => t.publicKey.toBase58() === nodeBankPK.toBase58());
            if (!nodeBankState)
                throw new Error('Node bank state not found');
            const accountData = [
                { pubkey: MangoHandler.MangoGrouPK, isWritable: true },
                { pubkey: mangoAccountPK, isWritable: true },
                { pubkey: mangoGroupState.mangoCache, isWritable: true },
                { pubkey: nodeBankPK, isWritable: true },
                { pubkey: nodeBankState.vault, isWritable: true },
                { pubkey: mangoGroupState.signerKey, isWritable: true },
                { pubkey: web3_js_1.PublicKey.default },
            ];
            const remainingAccounts = [];
            for (const account of accountData) {
                remainingAccounts.push({
                    pubkey: account.pubkey,
                    isWritable: !!account.isWritable,
                    isSigner: false,
                });
            }
            const [collateralVault] = yield web3_js_1.PublicKey.findProgramAddress([Buffer.from(constants_1.SEEDS.COLLATERAL_VAULT_PREFIX), new web3_js_1.PublicKey(strategy.pubkey).toBuffer()], program.programId);
            const txAccounts = {
                vault,
                strategy: strategy.pubkey,
                reserve: strategy.state.reserve,
                strategyProgram: MangoHandler.MangoProgramId,
                collateralVault,
                feeVault: vaultState.feeVault,
                tokenVault,
                userToken,
                userLp,
                tokenProgram: spl_token_1.TOKEN_PROGRAM_ID,
            };
            if (opt === null || opt === void 0 ? void 0 : opt.affiliate) {
                const tx = yield opt.affiliate.affiliateProgram.methods
                    .withdrawDirectlyFromStrategy(amount, new anchor.BN(0))
                    .preInstructions(preInstructions)
                    .postInstructions(postInstructions)
                    .remainingAccounts(remainingAccounts)
                    .accounts(Object.assign(Object.assign({}, txAccounts), { partner: opt.affiliate.partner, user: opt.affiliate.user, vaultProgram: program.programId, vaultLpMint: vaultState.lpMint, owner: walletPubKey }))
                    .transaction();
                return tx;
            }
            const tx = yield program.methods
                .withdrawDirectlyFromStrategy(amount, new anchor.BN(0))
                .preInstructions(preInstructions)
                .postInstructions(postInstructions)
                .remainingAccounts(remainingAccounts)
                .accounts(Object.assign(Object.assign({}, txAccounts), { lpMint: vaultState.lpMint, user: walletPubKey }))
                .transaction();
            return tx;
        });
    }
}
exports.default = MangoHandler;
MangoHandler.MangoProgramId = new web3_js_1.PublicKey('mv3ekLzLbnVPNxjSKvqBpU3ZeZXPQdEC3bp5MDEBG68');
MangoHandler.MangoGrouPK = new web3_js_1.PublicKey('98pjRuQjK3qA6gXts96PqZT4Ze5QmnCmt3QYjhbUSPue');
