// @mercurial-finance/vault-sdk/dist/cjs/src/vault/strategy/psylend.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const web3_js_1 = require("@solana/web3.js");
const psylend_utils_1 = require("@mithraic-labs/psylend-utils");
const spl_token_1 = require("@solana/spl-token");
const anchor_1 = require("@project-serum/anchor");
const constants_1 = require("../constants");
const utils_1 = require("../utils");
class PsyLendHandler {
    constructor(cluster, program) {
        this.cluster = cluster;
        this.connection = program.provider.connection;
        this.psyLendProgram = new anchor_1.Program(psylend_utils_1.IDL, psylend_utils_1.keys.psyLendMainnetProgramKey, new anchor_1.AnchorProvider(program.provider.connection, {}, anchor_1.AnchorProvider.defaultOptions()));
    }
    withdraw(walletPubKey, program, strategy, vault, tokenVault, vaultState, userToken, userLp, amount, preInstructions, postInstructions, opt) {
        return __awaiter(this, void 0, void 0, function* () {
            if (!walletPubKey)
                throw new Error('No user wallet public key');
            const strategyBuffer = new web3_js_1.PublicKey(strategy.pubkey).toBuffer();
            const [collateralVault] = web3_js_1.PublicKey.findProgramAddressSync([Buffer.from(constants_1.SEEDS.COLLATERAL_VAULT_PREFIX), strategyBuffer], program.programId);
            const { reserve } = strategy.state;
            const reserveState = yield this.psyLendProgram.account.reserve.fetchNullable(reserve);
            if (!reserveState)
                throw new Error('No user reserve account');
            const { market, depositNoteMint, pythOraclePrice, feeNoteVault, vault: reserveVault } = reserveState;
            const marketState = yield this.psyLendProgram.account.market.fetchNullable(market);
            if (!marketState)
                throw new Error('No user reserve market account');
            const { marketAuthority } = marketState;
            const [strategyOwnerPubkey] = web3_js_1.PublicKey.findProgramAddressSync([Buffer.from(constants_1.SEEDS.PSYLEND), strategyBuffer], program.programId);
            const [strategyOwnerAccount] = web3_js_1.PublicKey.findProgramAddressSync([Buffer.from(constants_1.SEEDS.PSYLEND_OWNER), reserve.toBuffer(), strategyOwnerPubkey.toBuffer()], psylend_utils_1.keys.psyLendMainnetProgramKey);
            const [tokenAccount, createTokenAccountIx] = yield (0, utils_1.getOrCreateATAInstruction)(vaultState.tokenMint, strategyOwnerPubkey, this.connection, {
                payer: walletPubKey,
            });
            createTokenAccountIx && preInstructions.push(createTokenAccountIx);
            const accounts = [
                { pubkey: market, isWritable: true },
                { pubkey: marketAuthority, isWritable: true },
                { pubkey: strategyOwnerAccount, isWritable: true },
                { pubkey: strategyOwnerPubkey, isWritable: true },
                { pubkey: reserveVault, isWritable: true },
                { pubkey: depositNoteMint, isWritable: true },
                { pubkey: tokenAccount, isWritable: true },
            ];
            const remainingAccounts = [];
            for (const account of accounts) {
                remainingAccounts.push({
                    pubkey: account.pubkey,
                    isWritable: true,
                    isSigner: false,
                });
            }
            const txAccounts = {
                vault,
                strategy: new web3_js_1.PublicKey(strategy.pubkey),
                reserve: new web3_js_1.PublicKey(strategy.state.reserve),
                strategyProgram: psylend_utils_1.keys.psyLendMainnetProgramKey,
                collateralVault,
                feeVault: vaultState.feeVault,
                tokenVault,
                userToken,
                userLp,
                tokenProgram: spl_token_1.TOKEN_PROGRAM_ID,
            };
            const accrueInterestIx = yield psylend_utils_1.instructions.accrueInterestIx(this.psyLendProgram, market, marketAuthority, reserve, feeNoteVault, depositNoteMint);
            preInstructions.push(accrueInterestIx);
            const refreshReserveIx = yield psylend_utils_1.instructions.refreshReserveIx(this.psyLendProgram, market, reserve, pythOraclePrice);
            preInstructions.push(refreshReserveIx);
            if (opt === null || opt === void 0 ? void 0 : opt.affiliate) {
                const tx = yield opt.affiliate.affiliateProgram.methods
                    .withdrawDirectlyFromStrategy(new anchor_1.BN(amount), new anchor_1.BN(0))
                    .accounts(Object.assign(Object.assign({}, txAccounts), { partner: opt.affiliate.partner, user: opt.affiliate.user, vaultProgram: program.programId, vaultLpMint: vaultState.lpMint, owner: walletPubKey }))
                    .remainingAccounts(remainingAccounts)
                    .preInstructions(preInstructions)
                    .postInstructions(postInstructions)
                    .transaction();
                return tx;
            }
            const tx = yield program.methods
                .withdrawDirectlyFromStrategy(new anchor_1.BN(amount), new anchor_1.BN(0))
                .accounts(Object.assign(Object.assign({}, txAccounts), { lpMint: vaultState.lpMint, user: walletPubKey }))
                .remainingAccounts(remainingAccounts)
                .preInstructions(preInstructions)
                .postInstructions(postInstructions)
                .transaction();
            return tx;
        });
    }
}
exports.default = PsyLendHandler;
