// @mercurial-finance/vault-sdk/dist/cjs/src/vault/strategy/solendWithoutLM.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const web3_js_1 = require("@solana/web3.js");
const solend = __importStar(require("@mercurial-finance/solend-sdk"));
const spl_token_1 = require("@solana/spl-token");
const anchor = __importStar(require("@project-serum/anchor"));
const constants_1 = require("../constants");
class SolendWithoutLMHandler {
    constructor(strategyProgram) {
        this.strategyProgram = strategyProgram;
    }
    getReserveState(program, reserve) {
        return __awaiter(this, void 0, void 0, function* () {
            const state = yield (() => __awaiter(this, void 0, void 0, function* () {
                const account = yield program.provider.connection.getAccountInfo(reserve);
                const solendParse = solend.parseReserve(account.owner, account);
                return solendParse.info;
            }))();
            return {
                collateral: {
                    mintPubkey: state.collateral.mintPubkey,
                    mintTotalSupply: state.collateral.mintTotalSupply.toNumber(),
                    supplyPubkey: state.collateral.mintTotalSupply.toString(),
                },
                state,
            };
        });
    }
    withdraw(walletPubKey, program, strategy, vault, tokenVault, vaultState, userToken, userLp, amount, preInstructions, postInstructions, opt) {
        return __awaiter(this, void 0, void 0, function* () {
            const { collateral, state } = yield this.getReserveState(program, strategy.state.reserve);
            let [collateralVault] = yield web3_js_1.PublicKey.findProgramAddress([Buffer.from(constants_1.SEEDS.COLLATERAL_VAULT_PREFIX), new web3_js_1.PublicKey(strategy.pubkey).toBuffer()], program.programId);
            const { liquidity, lendingMarket } = state;
            const [lendingMarketAuthority] = yield web3_js_1.PublicKey.findProgramAddress([lendingMarket.toBuffer()], this.strategyProgram);
            const accounts = [
                { pubkey: liquidity.supplyPubkey, isWritable: true },
                { pubkey: lendingMarket },
                { pubkey: lendingMarketAuthority },
                { pubkey: collateral.mintPubkey, isWritable: true },
                { pubkey: web3_js_1.SYSVAR_CLOCK_PUBKEY },
            ];
            const remainingAccounts = [];
            for (const account of accounts) {
                remainingAccounts.push({
                    pubkey: account.pubkey,
                    isWritable: !!account.isWritable,
                    isSigner: false,
                });
            }
            const { pythOracle, switchboardOracle } = liquidity;
            if (!pythOracle || !switchboardOracle) {
                throw Error('Cannot get pythOracle or switchboardOracle from Solend');
            }
            const txAccounts = {
                vault,
                strategy: strategy.pubkey,
                reserve: strategy.state.reserve,
                strategyProgram: this.strategyProgram,
                collateralVault,
                feeVault: vaultState.feeVault,
                tokenVault,
                userToken,
                userLp,
                tokenProgram: spl_token_1.TOKEN_PROGRAM_ID,
            };
            if (opt === null || opt === void 0 ? void 0 : opt.affiliate) {
                const tx = yield opt.affiliate.affiliateProgram.methods
                    .withdrawDirectlyFromStrategy(amount, new anchor.BN(0))
                    .accounts(Object.assign(Object.assign({}, txAccounts), { partner: opt.affiliate.partner, user: opt.affiliate.user, vaultProgram: program.programId, vaultLpMint: vaultState.lpMint, owner: walletPubKey }))
                    .remainingAccounts(remainingAccounts)
                    .preInstructions(preInstructions.concat([
                    solend.refreshReserveInstruction(strategy.state.reserve, this.strategyProgram, pythOracle, switchboardOracle),
                ]))
                    .postInstructions(postInstructions)
                    .transaction();
                return tx;
            }
            const tx = yield program.methods
                .withdrawDirectlyFromStrategy(amount, new anchor.BN(0))
                .accounts(Object.assign(Object.assign({}, txAccounts), { lpMint: vaultState.lpMint, user: walletPubKey }))
                .remainingAccounts(remainingAccounts)
                .preInstructions(preInstructions.concat([
                solend.refreshReserveInstruction(strategy.state.reserve, this.strategyProgram, pythOracle, switchboardOracle),
            ]))
                .postInstructions(postInstructions)
                .transaction();
            return tx;
        });
    }
}
exports.default = SolendWithoutLMHandler;
